---
title: "V3_stochastic"
output: html_document
---

```{r, results="hide"}
suppressPackageStartupMessages({
    library(Seurat)
    library(veloviz)
    library(reticulate)
    library(readxl)
    library(pheatmap)
    library(MAST)
    library(org.Pf.plasmo.db)
    library(clusterProfiler)
    library(cowplot)
    library(RColorBrewer)
    library(ggvenn)
    library(GenomicFeatures)
    library(rtracklayer)
    library(ggpubr)
    library(GGally)
    library("MetBrewer")
    library(tidyverse)
    library(viridis)
    library(SeuratDisk)
    library(dittoSeq)
  library(slingshot)
  library(tradeSeq)
    })
```

```{r setup}
knitr::opts_knit$set(root.dir = '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/sc_velo/')
```

```{r}
# rm(list=ls())
```


```{r}
.libPaths()
```

```{r}
use_condaenv("/Users/jr35/envs/scv_jupcon/", required = TRUE)
```

```{r}
scv <- reticulate::import("scvelo")
```

```{r}
scv$logging$print_version()
```

```{r}
ls()
```


```{r}
v3_bf2_names <- c("D5", "D10a","D10b", "D10a_D10b", "all_days")
v3_bf2_dnames <- c('all_days', 'd10_d10b', 'day10', 'day10b', 'day5')
```

```{r, eval=FALSE}
# ##Copy souporcell files for the optimum clusters
#   for (i in 1:length(v3_bf2_dnames)) {
#     system(
#       paste0('mkdir -p /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/PfDB52e/', v3_bf2_dnames[i])
#       )
#     system(
#       paste0('rsync -av jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/PfDB52e/', v3_bf2_dnames[i], '/scv_bf2_b4CR.h5ad /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/PfDB52e/', v3_bf2_dnames[i])
#     )
#   }
  
```


```{r, eval=FALSE}
# ##Copy souporcell files for the optimum clusters
#   v3_bf2_names <- c("D5", "D10a","D10b", "D10a_D10b", "all_days")
#   for (i in 1:length(sample_name)) {
#     system(
#       paste0('mkdir -p /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/raw/PfDB52e/', v3_bf2_names[i])
#       )
#     system(
#       paste0('rsync -av jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/sc_velo/data/raw/', sample_name[i], '/', source_irods[i], '_PfDB52e/filtered_feature_bc_matrix /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/raw/PfDB52e/', v3_bf2_names[i])
#     )
#   }
#   
```

```{r}
list.files("data/processed/PfDB52e", pattern = "scv_bf1_b4CR_stocha.h5ad", full.names = T, recursive = T, include.dirs = T)
```

```{r}
ef_scv_stochastic <- list.files("data/processed/PfDB52e", pattern = "scv_bf1_b4CR_stocha.h5ad", full.names = T, recursive = T, include.dirs = T) %>%
set_names(nm = (str_remove_all(.,'data/processed/PfDB52e/|/scv_bf1_b4CR_stocha.h5ad'))) %>%
map(scv$read) 
```

```{r}
ef_scv_stochastic$all_days$var <- ef_scv_stochastic$all_days$var%>% 
  mutate(Gene_ID = rownames(.))
ef_scv_stochastic$all$var <- ef_scv_stochastic$all$var%>%
  mutate(Gene_ID = rownames(.))

```


```{r}
# list.files("data/processed/PfDB52e", pattern = "scv_bf2_b4CR.h5ad", full.names = T, recursive = T, include.dirs = T)
```


```{r}
# ef_scv_stochastic <- list.files("data/processed/PfDB52e", pattern = "scv_bf2_b4CR.h5ad", full.names = T, recursive = T, include.dirs = T) %>%
# set_names(nm = (str_remove_all(.,'data/processed/PfDB52e/|/scv_bf2_b4CR.h5ad'))) %>%
# map(scv$read)
```

```{r}
# ef_scv_stochastic[[1]]
```

```{r}
# scv_dynamical_genes <- map(ef_scv_stochastic, ~scv$get_df(., 'rank_dynamical_genes/names'))
```

```{r, results="hide"}
# map(scv_dynamical_genes,  ~ head(.x[,c('gametocyte (male)','gametocyte (female)')] ))
```

```{r}
# png('plots/a.png')
# map2(ef_scv_stochastic,scv_dynamical_genes,  ~ scv$pl$scatter(.x, basis=.y[,'gametocyte (male)'][1:2] ), color= 'Stage',ncols=5, save = 'plots/a.png') 
# dev.off()
```

```{r}
# df = scv.get_df(obj, 'rank_dynamical_genes/names')
#     # print(df.head(5))
#     for cluster in ['gametocyte (female)', 'gametocyte (male)']:
#         scv.pl.scatter(obj, df[cluster][:5], ylabel=cluster, frameon=False, color= 'Stage', add_outline=cluster)
```

```{r}
# top_dynamical <- ef_scv_stochastic %>% 
# map(. %>%
#     .$var %>%
#     filter(fit_diff_kinetics %in% c('Gametocytes (early male)', 'Gametocytes (late male)', 'Gametocytes (early male),Gametocytes (late male)')) %>%
#     arrange(fit_pval_kinetics) %>%
#     dplyr::select('fit_diff_kinetics', 'fit_pval_kinetics','Chromosome') %>%
#     head() %>% 
#     rownames()) 
```

```{r}
# kwargs = dict(linewidth=2, add_linfit=True, frameon=False)
# for obj in ef_qcd_bf2_scv_varKine:
#     bf2_diffK = scv.get_df(obj, ['fit_diff_kinetics', 'fit_pval_kinetics','Chromosome'], precision=2)
#     top5_m_varkinetics = bf2_diffK[bf2_diffK.fit_diff_kinetics.isin(['Gametocytes (early male)', 'Gametocytes (late male)', 'Gametocytes (early male),Gametocytes (late male)'])].sort_values("fit_pval_kinetics").iloc[:5,].index.to_list()
#     scv.pl.scatter(obj, basis=top5_m_varkinetics, ncols=5, add_outline='fit_diff_kinetics', **kwargs, legend_loc='right')
```

```{r}
# top_kine_male <- ef_scv_stochastic %>% 
# map(. %>%
#     .$var %>%
#     filter(fit_diff_kinetics %in% c('Gametocytes (early male)', 'Gametocytes (late male)', 'Gametocytes (early male),Gametocytes (late male)')) %>%
#     arrange(fit_pval_kinetics) %>%
#     dplyr::select('fit_diff_kinetics', 'fit_pval_kinetics','Chromosome') %>%
#     head() %>% 
#     rownames()) 
```

```{r}
# png('plots/a.png')
# scv$pl$scatter(ef_scv_stochastic[[1]], basis=top_kine_male[[1]] , save = 'plots/a.png',color= 'Stage')
# dev.off()
```

```{r}
# png('plots/a.png')
# map2(ef_scv_stochastic,top_kine_male,  ~ scv$pl$scatter(.x, basis=.y ),ncols=5, save = 'plots/a.png') 
# dev.off()
```

```{r}
# ef_scv_stochastic %>% 
# map(. %>%
#     .$var %>%
#     filter(fit_diff_kinetics %in% c('Gametocytes (early male)', 'Gametocytes (late male)', 'Gametocytes (early male),Gametocytes (late male)')) %>%
#     arrange(fit_pval_kinetics) %>%
#     dplyr::select('fit_diff_kinetics', 'fit_pval_kinetics','Chromosome') %>%
#     head() %>% 
#     rownames()) 
```

```{r}
# ef_scv_stochastic <- list.files("data/processed/PfDB52e", pattern = "scv_bf2_b4CR.h5ad", full.names = T, recursive = T, include.dirs = T) %>%
# set_names(nm = (str_remove_all(.,'data/processed/PfDB52e/|/scv_bf2_b4CR.h5ad'))) %>%
# map(scv$read)
```

```{r}
scv_names <- names(ef_scv_stochastic)
```

```{r}
pf_introns <- read_csv('data/raw/introns.csv') %>% dplyr::select(-c(...1)) 
```

```{r}
gene_intron_size <-pf_introns %>% group_by(gene_id) %>% summarise(intron_size = sum(width), isoforms = n())
```

```{r}
# gene_intron_size 
```


```{r, fig.width= 28, fig.height=7}
# options(repr.plot.width=28, repr.plot.height=7)

ef_scv_stochastic %>% 
map(. %>%
    .$var %>%
    filter(velocity_genes =='TRUE') %>%
    dplyr::select('Gene_ID', 'total_unspliced', 'total_spliced') %>% 
    left_join(., gene_intron_size, by=c('Gene_ID'='gene_id'))) %>%
    map(., ~ggplot(.x, aes(x=log1p(intron_size),y=log1p(total_unspliced), color = isoforms)) +
        geom_point() + 
        geom_smooth(method='lm', formula= y~x)+ 
        stat_cor(method = "pearson", label.x = 4, label.y = 7.5,size = 6)+
theme(axis.text=element_text(size=18), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=20,face="bold"),
      strip.text.x = element_text(size = 20),
      strip.text.y = element_text(size = 20),
     plot.title = element_text(size = 20,hjust = 0.5,face="bold") ) + 
labs(x = "log1p(intron size (bases))", y = "log1p(total unspliced count)") +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="SexDE"))
        )%>%
plot_grid(plotlist = .,labels = names(ef_scv_stochastic), label_size = 24, hjust = -2.5,  ncol = 4)

# options(repr.plot.width=7, repr.plot.height=7)
```


```{r}
lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}  
```



```{r}
suppressWarnings({
for (i in c("total_spliced", "total_unspliced")) {
    print(GGally::ggpairs(ef_scv_stochastic %>%
                          map(. %>%
                              .$var %>%
                              dplyr::select('Gene_ID', all_of(i)))%>%
                          bind_rows(., .id = "Days") %>%
                          dplyr::select(c("Gene_ID", "Days", i))%>%
                          tidyr::spread(key = "Days", value = i) %>%
                          tibble::column_to_rownames("Gene_ID") %>%
                          dplyr::mutate_all(.funs = log1p),
                          lower = list(continuous = GGally::wrap(lowerfun)),
                          upper = list(continuous = wrap("cor", size = 7)),
                          title = paste0("Comparison of ", i," counts between days (datasets)"), 
                          xlab = "log(total gene count + 1)") +
        theme(axis.text=element_text(size=15), 
              axis.title=element_text(size=20,face="bold"), 
              legend.text = element_text(size=20), 
              legend.title = element_text(size=20,face="bold"),
                strip.text = element_text(size = 20),
             title = element_text(size = 13)) 
          )
    }
    })
```

```{r}
# + 
# rotate_x_text() +
# labs(x = "Sex DE", y = 'Counts') + 
# guides(fill=guide_legend(title="Sex correlation"))
```

```{r}
# suppressWarnings({
# for (i in c("fit_alpha", "fit_beta", "fit_gamma")) {
#     print(GGally::ggpairs(ef_scv_dyn %>%
#                           map(. %>%
#                               .$var %>%
#                               dplyr::select('Gene_ID', all_of(i)))%>%
#                           bind_rows(., .id = "Days") %>%
#                           dplyr::select(c("Gene_ID", "Days", i)) %>%
#                           tidyr::spread(key = "Days", value = i) %>%
#                           tibble::column_to_rownames("Gene_ID") %>%
#                           dplyr::mutate_all(.funs = log1p),
#                           lower = list(continuous = GGally::wrap(lowerfun)),
#                           upper = list(continuous = wrap("cor", size = 7)),
#                           title = paste0("Comparison of ", i," counts between days"), 
#                           xlab = paste("log(",i," + 1)")) +
#         theme(axis.text=element_text(size=15), 
#               axis.title=element_text(size=20,face="bold"), 
#               legend.text = element_text(size=20), 
#               legend.title = element_text(size=20,face="bold"),
#                 strip.text = element_text(size = 20),
#              title = element_text(size = 15)) +
#           rotate_x_text()
#           )
#     }
#     })
```



```{r}
#gene and cell names
bf2_genes <- ef_scv_stochastic %>% 
map(. %>%
    .$var %>% 
    dplyr::select(Gene_ID) %>% 
    rownames_to_column('bcode') %>% 
    mutate(across(Gene_ID, ~case_when(. == 'selfdefinedNA' ~ bcode,
                                        TRUE ~ as.character(.)))) %>% pull(Gene_ID)
    )

bf2_cells <- ef_scv_stochastic %>% 
map(. %>%
    .$obs_names %>%
    .$values
    )
```

```{r}
#get velocity for all qcd genes
bf2_vels <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
pmap(~ ..1$layers[['velocity']] %>%
    `colnames<-` (..2) %>%
    `rownames<-` (..3))
```

```{r}
# genes_w_velocities_gt0 <- bf2_vels %>%
# map(~.[,., na.rm=T)>0] )
```

```{r}
##Genes for which positive velocity was estimated succesful and the velocities are non-zero 

# options(repr.plot.width=7, repr.plot.height=7)
bf2_vels %>%
map(~.[,colSums(., na.rm=T)!=0] %>% 
    colnames
    ) %>% 
# .[2:5] %>%
ggvenn(.,fill_color = c("#0073C2FF", "#EFC000FF"),
       stroke_size = 0.5, set_name_size = 5
      )
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
##Genes estimated by the steady state model and used as input for the dynamical model for the fitting 
ef_scv_stochastic %>% 
map(. %>%
    .$var %>%
    filter(velocity_genes == TRUE) %>%
    dplyr::select(Gene_ID, velocity_genes) %>% 
    pull(Gene_ID)
    ) %>% 
# .[2:5] %>%
ggvenn(.,fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868", "#A67438"),
       stroke_size = 0.5, set_name_size = 5
      )

# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
list(ef_scv_stochastic,bf2_vels)  %>% 
pmap(~ ..1$var %>%
    dplyr::select(Gene_ID, velocity_genes) %>%
    left_join(., colSums(..2) %>% data.frame() %>%rename('velo_sums'='.') %>%rownames_to_column('bcode'), by=c('Gene_ID'='bcode') ) %>%
    mutate(`Velocities estimated` = ifelse(is.na(velo_sums), 'No', 'Yes')) %>%
     # filter(!is.na(velo_sums)) %>%
    dplyr::count(`Velocities estimated`)
    ) %>%
bind_rows(., .id='Days') %>%
ggplot(., aes(x=Days, y=n, fill=`Velocities estimated`)) + 
geom_col(position = position_dodge2(width = 0.2)) +
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =1.2, size =8, angle = 90)+
theme(axis.text=element_text(size=18), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=20,face="bold"),
      strip.text.x = element_text(size = 20),
      strip.text.y = element_text(size = 20),
     plot.title = element_text(size = 20,hjust = 0.5,face="bold") ) + 
rotate_x_text() +
guides(fill=guide_legend(title="Velocities \nestimated")) +
labs(y='Count')
```

```{r}
list(ef_scv_stochastic,bf2_vels)  %>% 
pmap(~ ..1$var %>%
    dplyr::select(Gene_ID, velocity_genes,total_unspliced) %>%
    left_join(., colSums(..2) %>% data.frame() %>%rename('velo_sums'='.') %>%rownames_to_column('bcode'), by=c('Gene_ID'='bcode') ) %>%
    mutate(`Velocities estimated` = ifelse(is.na(velo_sums), 'No', 'Yes'))# %>%
     # filter(!is.na(velo_sums)) %>%
    # count(`Velocities estimated`)
    ) %>%
bind_rows(., .id='Days') %>%
ggplot(., aes(x=`Velocities estimated`, y=log1p(total_unspliced), fill=Days)) + 
geom_boxplot()
```




```{r, eval=FALSE}
system(
  paste0('rsync -av --include="*drivers_in*" --exclude="*" jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/* /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/RoadtoUMAP/')
  )
```




```{r}
stocha_corr_genes <- list.files("data/processed/RoadtoUMAP/", pattern = "_PfDB_stocha.csv", full.names = T, recursive = T, include.dirs = T) %>%
set_names(nm = (str_remove_all(.,'data/processed/RoadtoUMAP//|_PfDB_stocha.csv'))) %>%
map(read_csv) %>%
map(. %>%
    rename('ID' = 1)%>%
    rename_at(vars(matches('^Gametocyte|^Early|^Female|^Male')), ~ str_remove(.x,'Gametocyte.|Early schizont.|Female.|Male.')))
```


```{r}
list.files("data/processed/RoadtoUMAP/", pattern = "_vkck|_pk", full.names = T, recursive = T, include.dirs = T)%>%
set_names(nm = (str_remove_all(.,'data/processed/RoadtoUMAP//|_PfDB|.csv|_bf1|_bf2')))
```

```{r}
stocha_pk_corr_genes <- list.files("data/processed/RoadtoUMAP/", pattern = "_vkck|_pk", full.names = T, recursive = T, include.dirs = T)%>%
set_names(nm = (str_remove_all(.,'data/processed/RoadtoUMAP//|_PfDB|.csv|_bf1|_bf2'))) %>%
map(read_csv) %>%
map(. %>%
    rename('ID' = 1)%>%
    rename_at(vars(matches('^Gametocyte|^Early|^Female|^Male|^Asexual')), ~ str_remove(.x,'Gametocyte.|Early schizont.|Female.|Male.|Asexual.')))
```

```{r}
# corr_genes <- list.files("data/processed/RoadtoUMAP/", pattern = "_PfDB.csv", full.names = T, recursive = T, include.dirs = T) %>%
# set_names(nm = (str_remove_all(.,'data/processed/RoadtoUMAP//|_PfDB.csv'))) %>%
# map(read_csv) %>%
# map(. %>%
#     rename('ID' = 1)%>%
#     rename_at(vars(matches('^Female|^Male')), ~ str_remove(.x,'Female.|Male.')))
```

```{r}
stocha_pk_corr_genes %>%
map(. %>%
    dplyr::filter(qval < 0.00005 & (corr > 0.4|corr < -0.4)) %>%
    arrange(desc(corr)) %>%
   pull(`Gene`))
```

```{r}
early_gam <- c('PF3D7_1477300','PF3D7_0936600', 'PF3D7_1016900', 'PF3D7_0406200')
```

```{r}
andy_markers <- c('PF3D7_0603600','PF3D7_1438800', 'PF3D7_1319600', 'PF3D7_1241400','PF3D7_0927200', 'PF3D7_0414500', 'PF3D7_1220000')
```

```{r}
juli_markers<- c('PF3D7_1466800','PF3D7_1146800', 'PF3D7_0114000', 'PF3D7_1038400', 'PF3D7_0608800', 'PF3D7_1216500', 'PF3D7_0209000', 'PF3D7_1103800')
```

```{r}
stocha_pk_corr_genes %>% 
map(. %>%
    dplyr::filter(qval < 0.005 & ( corr < -0.2 | corr > 0.2)) %>%
    dplyr::filter(ID %in% juli_markers) %>%
    arrange(desc(corr)) )
```

```{r}
stocha_pk_corr_genes %>% 
map(. %>%
    dplyr::filter(qval < 0.005 & ( corr < -0.2 | corr > 0.2)) %>%
    dplyr::filter(ID %in% andy_markers) %>%
    arrange(desc(corr)) )
```

```{r}
stocha_pk_corr_genes %>% 
map(. %>%
    dplyr::filter(qval < 0.005 & ( corr < -0.2 | corr > 0.2)) %>%
    dplyr::filter(ID %in% early_gam) %>%
    arrange(desc(corr)) )
```


```{r}
Road2UMAP <- readRDS("data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.2.RDS")
```

```{r}
##Save object out for Juli as is
# SaveH5Seurat(Road2UMAP, filename = "Road2UMAP.h5Seurat")
```

```{r}
##Save object out for Juli as is
# Convert("Road2UMAP.h5Seurat", dest = "h5ad")
```

```{r}
dittoBarPlot(
  object = Road2UMAP,
  var = 'Stage',
  group.by = "orig.ident",
  color.panel = rev(c("#CAE69F", "#78C679",  '#FEEEAA', '#FEB24C','#C9E8F1', '#85B1D3', 'thistle',  'mediumpurple',  '#551A8B')),
  var.labels.reorder = rev(c(1,7,3,9, 2, 8,4,5,6)),
  x.reorder =  c(1,4,5,2,3),
  main = NULL,
  xlab = NULL,
  ylab = "Fraction of cells",
  scale = "percent"
)
```

```{r}
Road2UMAP@meta.data <- Road2UMAP@meta.data %>% mutate(across(where(is.factor), as.character)) 
```

```{r}
DimPlot(Road2UMAP, reduction = "umap", dims = c(2,3))
```

```{r}
FeaturePlot(Road2UMAP, features = c('PF3D7-1348800', 'PF3D7-1477700','PF3D7-0500800', 'PF3D7-0114000', 'PF3D7-0220200'), dims = c(2,3))
```


```{r}
stocha_gns <- stocha_corr_genes %>% 
  map(
    . %>% select(ID, Timepoint_freq, Gene, source_id, `Gene ID`, qval, corr,Timepoint) %>%
      dplyr::filter(qval < 0.001 )%>%
      dplyr::filter(corr < -0.2 | corr > 0.2) %>%
      mutate(dir = case_when(corr<0 ~ 'anti-correlated',
                             corr>0 ~ 'correlated')) %>%
      pivot_wider(names_from = Timepoint, values_from = c(corr, qval,dir))
    ) %>%
  bind_rows(., .id='sex_n_cluster') %>%
  pivot_wider(names_from = 'sex_n_cluster', values_from = c('corr_all_days', 'corr_all', 'qval_all_days', 'qval_all', 'dir_all_days', 'dir_all')) %>%
  mutate(
    CDgam_sex = case_when(
      if_any(matches('di.*_female_drivers_in_CDB'), ~.== 'correlated') & if_any(matches('di.*_male_drivers_in_CDB'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('di.*_female_drivers_in_CDB'), ~.== 'correlated') ~ 'Female',
      if_any(matches('di.*_male_drivers_in_CDB'), ~.== 'correlated') ~ 'Male'
      ),
    Bgam_sex = case_when(
      if_any(matches('di.*_female_drivers_in_gambranch'), ~.== 'correlated') & if_any(matches('di.*_male_drivers_in_gambranch'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('di.*_female_drivers_in_gambranch'), ~.== 'correlated') ~'Female',
      if_any(matches('di.*_male_drivers_in_gambranch'), ~.== 'correlated') ~ 'Male'
      ),
    lR_eT_Asex_gam = case_when(
      if_any(matches('di.*_early_schizont_drivers_in_lReT'), ~.== 'correlated') & if_any(matches('di.*_early_gam_drivers_in_lReT'), ~.== 'correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('di.*_early_schizont_drivers_in_lReT'), ~.== 'correlated')  ~ 'Asexual',
      if_any(matches('di.*_early_gam_drivers_in_lReT'), ~.== 'correlated') ~ 'Gametocyte'
      ),
    CDgam_sex_anti = case_when(
      if_any(matches('di.*_female_drivers_in_CDB'), ~.== 'anti-correlated') & if_any(matches('di.*_male_drivers_in_CDB'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('di.*_female_drivers_in_CDB'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('di.*_male_drivers_in_CDB'), ~.== 'anti-correlated')  ~ 'Male'
      ),
    Bgam_sex_anti = case_when(
      if_any(matches('di.*_female_drivers_in_gambranch'), ~.== 'anti-correlated') & if_any(matches('di.*_male_drivers_in_gambranch'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('di.*_female_drivers_in_gambranch'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('di.*_male_drivers_in_gambranch'), ~.== 'anti-correlated') ~ 'Male'
      ),
    lR_eT_Asex_gam_anti = case_when(
      if_any(matches('di.*_early_schizont_drivers_in_lReT'), ~.== 'anti-correlated') & if_any(matches('di.*_early_gam_drivers_in_lReT'), ~.== 'anti-correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('di.*_early_schizont_drivers_in_lReT'), ~.== 'anti-correlated') ~ 'Asexual',
      if_any(matches('di.*_early_gam_drivers_in_lReT'), ~.== 'anti-correlated')  ~ 'Gametocyte'
      ),
    CDgam_sex_all_days = case_when(
      if_any(matches('dir_all_d.*_female_drivers_in_CDB'), ~.== 'correlated') & if_any(matches('dir_all_d.*_male_drivers_in_CDB'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('dir_all_d.*_female_drivers_in_CDB'), ~.== 'correlated')  ~ 'Female',
      if_any(matches('dir_all_d.*_male_drivers_in_CDB'), ~.== 'correlated')  ~ 'Male'
      ),
    Bgam_sex_all_days = case_when(
      if_any(matches('dir_all_d.*_female_drivers_in_gambranch'), ~.== 'correlated') & if_any(matches('dir_all_d.*_male_drivers_in_gambranch'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('dir_all_d.*_female_drivers_in_gambranch'), ~.== 'correlated')  ~ 'Female',
      if_any(matches('dir_all_d.*_male_drivers_in_gambranch'), ~.== 'correlated')  ~ 'Male'
      ),
    lR_eT_Asex_gam_all_days = case_when(
      if_any(matches('dir_all_d.*_early_schizont_drivers_in_lReT'), ~.== 'correlated') & if_any(matches('dir_all_d.*_early_gam_drivers_in_lReT'), ~.== 'correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('dir_all_d.*_early_schizont_drivers_in_lReT'), ~.== 'correlated')  ~ 'Asexual',
      if_any(matches('dir_all_d.*_early_gam_drivers_in_lReT'), ~.== 'correlated')  ~ 'Gametocyte'
      ),
    CDgam_sex_all_days_anti = case_when(
      if_any(matches('dir_all_d.*_female_drivers_in_CDB'), ~.== 'anti-correlated') & if_any(matches('dir_all_d.*_male_drivers_in_CDB'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('dir_all_d.*_female_drivers_in_CDB'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('dir_all_d.*_male_drivers_in_CDB'), ~.== 'anti-correlated')  ~ 'Male'
      ),
    Bgam_sex_all_days_anti = case_when(
      if_any(matches('dir_all_d.*_female_drivers_in_gambranch'), ~.== 'anti-correlated') & if_any(matches('dir_all_d.*_male_drivers_in_gambranch'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('dir_all_d.*_female_drivers_in_gambranch'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('dir_all_d.*_male_drivers_in_gambranch'), ~.== 'anti-correlated')  ~ 'Male'
      ),
    lR_eT_Asex_gam_all_days_anti = case_when(
      if_any(matches('dir_all_d.*_early_schizont_drivers_in_lReT'), ~.== 'anti-correlated') & if_any(matches('dir_all_d.*_early_gam_drivers_in_lReT'), ~.== 'anti-correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('dir_all_d.*_early_schizont_drivers_in_lReT'), ~.== 'anti-correlated')  ~ 'Asexual',
      if_any(matches('dir_all_d.*_early_gam_drivers_in_lReT'), ~.== 'anti-correlated')  ~ 'Gametocyte'
      )
          ) %>%
  dplyr::rename_all(~str_replace_all(.,"male_drivers", "male")) %>%
  dplyr::rename_all(~str_replace_all(.,"early_schizont_drivers", "asexual")) %>%
  dplyr::rename_all(~str_replace_all(.,"early_gam_drivers", "gametocyte"))
  
```



```{r}
##Help with map https://stackoverflow.com/questions/68231455/how-to-use-map-if-based-on-the-name-of-each-element-in-the-list
# stocha_pk_gns <- stocha_pk_corr_genes %>%
#     map2(names(stocha_pk_corr_genes), ~ if(.y %in% c("early_gam_drivers_in_lReT_stocha", "early_schizont_drivers_in_lReT_stocha", "female_drivers_in_CDB_stocha", "female_drivers_in_gambranch_stocha", "male_drivers_in_CDB_stocha","male_drivers_in_gambranch_stocha")) {
#         .x %>% 
#             filter(Timepoint == 'all_days')%>% 
#             select(ID, Gene, source_id, `Gene ID`, qval, corr) %>%
#             head()
#     } else {
#         .x %>% 
#             select(ID, Gene, source_id, `Gene ID`, qval, corr)
#     }) %>%map(. %>%
#     dplyr::filter(qval < 0.001 )%>%
#     dplyr::filter(corr < -0.2 | corr > 0.2) %>%
#     mutate(dir = case_when(corr<0 ~ 'anti-correlated',
#                            corr>0 ~ 'correlated'))) %>%
stocha_pk_gns <- stocha_pk_corr_genes %>% 
  map(
    . %>% select(ID, Gene, source_id, `Gene ID`, qval, corr) %>%
      dplyr::filter(qval < 0.001 )%>%
      dplyr::filter(corr < -0.15 | corr > 0.15) %>%
      mutate(dir = case_when(corr<0 ~ 'anti-correlated',
                             corr>0 ~ 'correlated'))) %>%
  bind_rows(., .id='sex_n_cluster') %>%
  pivot_wider(names_from = 'sex_n_cluster', values_from = c('corr', 'qval', 'dir')) %>%
  mutate(
    Bgam_sex_stocha_VKCK = case_when(
      if_any(matches('d.*_female_drivers_vkck'), ~.== 'correlated') & if_any(matches('d.*_male_drivers_vkck'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('d.*_female_drivers_vkck'), ~.== 'correlated') ~'Female',
      if_any(matches('di.*_male_drivers_vkck'), ~.== 'correlated') ~ 'Male'
      ),
    Asex_gam_stocha_VKCK = case_when(
      if_any(matches('d.*_asexual_drivers_vkck'), ~.== 'correlated') & if_any(matches('d.*_early_gam_drivers_vkck'), ~.== 'correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('d.*_asexual_drivers_vkck'), ~.== 'correlated')  ~ 'Asexual',
      if_any(matches('d.*_early_gam_drivers_vkck'), ~.== 'correlated') ~ 'Gametocyte'
      ),
    Bgam_sex_stocha_VKCK_anti = case_when(
      if_any(matches('d.*_female_drivers_vkck'), ~.== 'anti-correlated') & if_any(matches('d.*_male_drivers_vkck'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('d.*_female_drivers_vkck'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('d.*_male_drivers_vkck'), ~.== 'anti-correlated') ~ 'Male'
      ),
    Asex_gam_stocha_VKCK_anti = case_when(
      if_any(matches('d.*_asexual_drivers_vkck'), ~.== 'anti-correlated') & if_any(matches('d.*_early_gam_drivers_vkck'), ~.== 'anti-correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('d.*_asexual_drivers_vkck'), ~.== 'anti-correlated') ~ 'Asexual',
      if_any(matches('d.*_early_gam_drivers_vkck'), ~.== 'anti-correlated')  ~ 'Gametocyte'
      ),
    Bgam_sex_PK = case_when(
      if_any(matches('d.*_female_drivers_pk'), ~.== 'correlated') & if_any(matches('d.*_male_drivers_pk'), ~.== 'correlated') ~ 'Female,Male',
      if_any(matches('d.*_female_drivers_pk'), ~.== 'correlated')  ~ 'Female',
      if_any(matches('d.*_male_drivers_pk'), ~.== 'correlated')  ~ 'Male'
      ),
    Asex_gam_PK = case_when(
      if_any(matches('d.*_asexual_drivers_pk'), ~.== 'correlated') & if_any(matches('d.*_early_gam_drivers_pk'), ~.== 'correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('d.*_asexual_drivers_pk'), ~.== 'correlated')  ~ 'Asexual',
      if_any(matches('d.*_early_gam_drivers_pk'), ~.== 'correlated')  ~ 'Gametocyte'
      ),
    Bgam_sex_PK_anti = case_when(
      if_any(matches('d.*_female_drivers_pk'), ~.== 'anti-correlated') & if_any(matches('d.*_male_drivers_pk'), ~.== 'anti-correlated') ~ 'Female,Male',
      if_any(matches('d.*_female_drivers_pk'), ~.== 'anti-correlated')  ~ 'Female',
      if_any(matches('d.*_male_drivers_pk'), ~.== 'anti-correlated')  ~ 'Male'
      ),
    Asex_gam_PK_anti = case_when(
      if_any(matches('d.*_asexual_drivers_pk'), ~.== 'anti-correlated') & if_any(matches('d.*_early_gam_drivers_pk'), ~.== 'anti-correlated') ~ 'Asexual,Gametocyte',
      if_any(matches('d.*_asexual_drivers_pk'), ~.== 'anti-correlated')  ~ 'Asexual',
      if_any(matches('d.*_early_gam_drivers_pk'), ~.== 'anti-correlated')  ~ 'Gametocyte'
      )
          ) %>%
  dplyr::rename_all(~str_replace_all(.,"male_drivers", "male")) %>%
  dplyr::rename_all(~str_replace_all(.,"early_schizont_drivers", "asexual")) %>%
  dplyr::rename_all(~str_replace_all(.,"early_gam_drivers", "gametocyte"))%>%
  dplyr::rename_all(~str_replace_all(.,"asexual_drivers", "asexual"))
  

##Assesment 
##stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)&!is.na(Bgam_sex_stocha_VKCK_anti)&!is.na(dir_male_pk)) %>% select(Bgam_sex_PK,Bgam_sex_PK_anti, Bgam_sex_stocha_VKCK, Bgam_sex_stocha_VKCK_anti, dir_female_pk, dir_male_pk, dir_female_vkck, dir_male_vkck) %>% count(dir_female_pk, dir_male_pk, dir_female_vkck, dir_male_vkck)
```

```{r}
print(GGally::ggpairs(stocha_pk_gns %>%
                          filter(!is.na(Bgam_sex_PK_anti) & !is.na(Bgam_sex_PK)) %>%
                            dplyr::select(corr_male_vkck, corr_female_vkck,corr_male_pk, corr_female_pk, ID) %>%
                            tibble::column_to_rownames("ID"),
                          # dplyr::mutate_all(.funs = log1p),
                          lower = list(continuous = GGally::wrap(lowerfun)),
                          title = paste0("Comparison of correlations counts between sex and driver genes methods")) +
          theme(strip.text = element_text(size = 8.75 - 0.25 * 1))
          )
```



```{r}
##regex https://stackoverflow.com/questions/23518325/how-to-extract-substring-between-patterns-and-in-r 
corr_clustr_venn <- function(dset= stocha_gns, corr_stage1 = 'dir_a.*_female_in_CDB', corr_stage2 = 'dir_a.*_male_in_CDB', sufx='ness'){

lst<- list(
  dset %>% filter(if_any(matches(corr_stage1), ~.== 'anti-correlated') ) %>% pull(ID),
  dset %>% filter(if_any(matches(corr_stage2), ~.== 'correlated') ) %>% pull(ID),
  dset %>% filter(if_any(matches(corr_stage1), ~.== 'correlated') ) %>% pull(ID),
  dset %>% filter(if_any(matches(corr_stage2), ~.== 'anti-correlated') ) %>% pull(ID)
  )
  names(lst) =str_to_sentence(
    c( paste0(str_extract(corr_stage1,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', 'anti-correlated'),
       paste0(str_extract(corr_stage2,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', 'correlated') ,
       paste0(str_extract(corr_stage1,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', 'correlated'),
       paste0(str_extract(corr_stage2,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', 'anti-correlated'))
  )
  ggvenn(lst,
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, 
       set_name_size = 6, 
       text_size = 5)+ 
scale_x_continuous(expand = expansion(mult = .25))+ 
scale_y_continuous(expand = expansion(mult = .15))
}
corr_clustr_venn()
```


```{r}
##BF1
##Assess the overlap between genes based on correlation (or anti-correlation) in CDB to either femaleness or maleness in overal(all) and alldays
##All
corr_clustr_venn(corr_stage1 = 'dir_.*_asexual_in_lReT', corr_stage2 = 'dir_.*_gametocyte_in_lReT', sufx='')

##All_days  independently
corr_clustr_venn(corr_stage1 = 'dir_all_d.*_asexual_in_lReT', corr_stage2 = 'dir_all_d.*_gametocyte_in_lReT', sufx='')
```

```{r}
##BF2
##All & All DAYS CDB and gambranch correlation and (anti-)correlation
corr_clustr_venn(corr_stage1 = 'dir_a.*_female_in', corr_stage2 = 'dir_a.*_male_in', sufx='ness')

##All DAYS CDB & gambranch 
corr_clustr_venn(corr_stage1 = 'dir_all_da.*_female_in', corr_stage2 = 'dir_all_da.*_male_in', sufx='ness')

##All DAYS CDB independently
corr_clustr_venn(corr_stage1 = 'dir_all_da.*_female_in_CDB', corr_stage2 = 'dir_all_da.*_male_in_CDB', sufx='ness')

##All DAYS gambranch independently
corr_clustr_venn(corr_stage1 = 'dir_all_da.*_female_in_gambranch', corr_stage2 = 'dir_all_da.*_male_in_gambranch', sufx='ness')
```


```{r}
###BF1
#VKCK & PK male and femaleness correlation and (anti-)correlation
corr_clustr_venn(dset = stocha_pk_gns, corr_stage1 = 'd.*_asexual_' , corr_stage2 = 'd.*_gametocyte_', sufx='ness')

##PK correlation and (anti-)correlation
corr_clustr_venn(dset=stocha_pk_gns, corr_stage1 = 'd.*_asexual_pk', corr_stage2 = 'd.*_gametocyte_pk', sufx='ness')
corr_clustr_venn(dset=stocha_pk_gns, corr_stage1 = 'd.*_asexual_vkck', corr_stage2 = 'd.*_gametocyte_vkck', sufx='ness')

```

```{r}
###BF2
#VKCK & PK male and femaleness correlation and (anti-)correlation
corr_clustr_venn(dset = stocha_pk_gns, corr_stage1 = 'd.*_female_' , corr_stage2 = 'd.*_male_', sufx='ness')

##PK correlation and (anti-)correlation
corr_clustr_venn(dset=stocha_pk_gns, corr_stage1 = 'd.*_female_pk', corr_stage2 = 'd.*_male_pk', sufx='ness')
corr_clustr_venn(dset=stocha_pk_gns, corr_stage1 = 'd.*_female_vkck', corr_stage2 = 'd.*_male_vkck', sufx='ness')

```


```{r}
# stocha_gns %>% filter(!is.na(Bgam_sex_all_days) & !is.na(Bgam_sex_all_days_anti)) %>% select(ID, Gene, Bgam_sex_all_days, Bgam_sex_all_days_anti, corr_all_days_male_in_gambranch, qval_all_days_male_in_gambranch, corr_all_days_female_in_gambranch, qval_all_days_female_in_gambranch) %>% View
```

```{r}
# stocha_gns %>% filter(!is.na(lR_eT_Asex_gam_all_days) | !is.na(lR_eT_Asex_gam_all_days_anti)) %>% select(ID,Gene, lR_eT_Asex_gam_all_days, lR_eT_Asex_gam_all_days_anti, corr_all_days_gametocyte_in_lReT , qval_all_days_gametocyte_in_lReT,corr_all_days_asexual_in_lReT , qval_all_days_asexual_in_lReT) %>% View
```



```{r}
##BF1&2 combi VS BF2
##regex https://stackoverflow.com/questions/23518325/how-to-extract-substring-between-patterns-and-in-r 
corr_clustr_venn_flex <- function(tbl1, tbl2, tbl3, tbl4, dir1, dir2, dir3, dir4, corr_stage1, corr_stage2, corr_stage3, corr_stage4, sufx='ness',prefx ){

lst<- list(
  tbl1 %>% filter(if_any(matches(corr_stage1), ~.== dir1) ) %>% pull(ID),
  tbl2 %>% filter(if_any(matches(corr_stage2), ~.== dir2) ) %>% pull(ID),
  tbl3 %>% filter(if_any(matches(corr_stage3), ~.== dir3) ) %>% pull(ID),
  tbl4 %>% filter(if_any(matches(corr_stage4), ~.== dir4) ) %>% pull(ID)
  )
  names(lst) =str_to_sentence(
    c( paste0(str_extract(corr_stage1,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', dir1),
       paste0(prefx,str_extract(corr_stage2,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', dir2),
       paste0(str_extract(corr_stage3,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', dir3) ,
       paste0(prefx,str_extract(corr_stage4,"(?<=(\\*l_|\\*_))(.+)(?=_in|_vkck|_pk|_$)"),sufx,'\n', dir4))
  )
  ggvenn(lst,
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4

      )
}
```




```{r}
##BF2
##PK vs stocha  corr
corr_clustr_venn_flex(tbl1 = stocha_pk_gns, 
                 tbl2 = stocha_pk_gns, 
                 tbl3 = stocha_pk_gns, 
                 tbl4 = stocha_pk_gns, 
                 dir1 = 'correlated', 
                 dir2 = 'correlated', 
                 dir3 = 'correlated', 
                 dir4 = 'correlated', 
                 corr_stage1 = 'd.*_female_vkck', 
                 corr_stage2 = 'd.*_female_pk', 
                 corr_stage3 = 'd.*_male_vkck', 
                 corr_stage4 ='d.*_male_pk', 
                 sufx='ness',
                 prefx='PK ')

##PK vs stocha CDB anti-corr
corr_clustr_venn_flex(tbl1 = stocha_pk_gns, 
                 tbl2 = stocha_pk_gns, 
                 tbl3 = stocha_pk_gns, 
                 tbl4 = stocha_pk_gns, 
                 dir1 = 'anti-correlated', 
                 dir2 = 'anti-correlated', 
                 dir3 = 'anti-correlated', 
                 dir4 = 'anti-correlated', 
                 corr_stage1 = 'd.*_female_vkck', 
                 corr_stage2 = 'd.*_female_pk', 
                 corr_stage3 = 'd.*_male_vkck', 
                 corr_stage4 ='d.*_male_pk', 
                 sufx='ness',
                 prefx='PK ')


```



```{r}
##BF1
##PK vs stocha corr cells correlation only
corr_clustr_venn_flex(tbl1 = stocha_pk_gns, 
                 tbl2 = stocha_pk_gns, 
                 tbl3 = stocha_pk_gns, 
                 tbl4 = stocha_pk_gns, 
                 dir1 = 'correlated', 
                 dir2 = 'correlated', 
                 dir3 = 'correlated', 
                 dir4 = 'correlated', 
                 corr_stage1 = 'd.*_gametocyte_vkck', 
                 corr_stage2 = 'd.*_gametocyte_pk', 
                 corr_stage3 = 'd.*_asexual_vkck', 
                 corr_stage4 ='d.*_asexual_pk', 
                 sufx='ness',
                 prefx='PK ')

##PK vs stocha  anti-corr cells correlation only
corr_clustr_venn_flex(tbl1 = stocha_pk_gns, 
                 tbl2 = stocha_pk_gns, 
                 tbl3 = stocha_pk_gns, 
                 tbl4 = stocha_pk_gns, 
                 dir1 = 'anti-correlated', 
                 dir2 = 'anti-correlated', 
                 dir3 = 'anti-correlated', 
                 dir4 = 'anti-correlated', 
                 corr_stage1 = 'd.*_gametocyte_vkck', 
                 corr_stage2 = 'd.*_gametocyte_pk', 
                 corr_stage3 = 'd.*_asexual_vkck', 
                 corr_stage4 ='d.*_asexual_pk', 
                 sufx='ness',
                 prefx='PK ')
```


```{r}
##BF1&2
##All DAYS anti-corr
corr_clustr_venn_flex(tbl1 = stocha_gns, 
                 tbl2 = stocha_gns, 
                 tbl3 = stocha_gns, 
                 tbl4 = stocha_gns, 
                 dir1 = 'anti-correlated', 
                 dir2 = 'anti-correlated', 
                 dir3 = 'anti-correlated', 
                 dir4 = 'anti-correlated', 
                 corr_stage1 = 'dir_all_d.*_female_in_gambranch', 
                 corr_stage2 = 'dir_all_d.*_male_in_gambranch', 
                 corr_stage3 = 'dir_all_d.*_asexual_in', 
                 corr_stage4 ='dir_all_d.*_gametocyte_in', 
                 sufx='ness',
                 prefx='')

##All DAYS corr
corr_clustr_venn_flex(tbl1 = stocha_gns, 
                 tbl2 = stocha_gns, 
                 tbl3 = stocha_gns, 
                 tbl4 = stocha_gns, 
                 dir1 = 'correlated', 
                 dir2 = 'correlated', 
                 dir3 = 'correlated', 
                 dir4 = 'correlated', 
                 corr_stage1 = 'dir_all_d.*_female_in_gambranch', 
                 corr_stage2 = 'dir_all_d.*_male_in_gambranch', 
                 corr_stage3 = 'dir_all_d.*_asexual_in', 
                 corr_stage4 ='dir_all_d.*_gametocyte_in', 
                 sufx='ness',
                 prefx='')
```


```{r}
api_ap2 <- c('PF3D7_1222600','PF3D7_1429200', 'PF3D7_0404100','PF3D7_1350900','PF3D7_1449500','PF3D7_0802100','AP2-G','PF3D7_0611200','PF3D7_1408200','PF3D7_1317200','PF3D7_1222400','PF3D7_0934400','PF3D7_0516800','PF3D7_1143100','PF3D7_1239200','PF3D7_0613800','PF3D7_1305200','PF3D7_1342900','PF3D7_1456000','PF3D7_0420300','PF3D7_1107800','PF3D7_1139300','PF3D7_1007700','PF3D7_1115500','PF3D7_1466400','PF3D7_0622900','PF3D7_0604100','PF3D7_0703700')
```

```{r}
stocha_pk_gns%>% filter(ID %in% api_ap2)
```

```{r}
candidate_published_gns <- toupper(c('PF3D7_1247800', 'PF3D7_1401900','PF3D7_0404100', 'PF3D7_0516800', 'PF3D7_1429200', 'PF3D7_1315800', 'PF3D7_0603600', 'PF3D7_1241400', 'PF3D7_0715100', 'pf3d7_1250100', 'pf3d7_1031000', 'pf3d7_1475500', 'pf3d7_1455800', 'pf3d7_1407000', 'pf3d7_0903800', 'pf3d7_0918700', 'pf3d7_0730400', 'pf3d7_1007800', 'pf3d7_1466800', 'pf3d7_1146800', 'pf3d7_1214800'))
```

```{r}
stocha_pk_gns %>% filter(ID %in% candidate_published_gns)
```

```{r, eval=FALSE}
system(
  paste0('rsync -av jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/raw/published /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/raw/')
  )
```

```{r}
VBcomplete_transcriptome <- readxl::read_excel("data/raw/published/12864_2019_6322_MOESM1_ESM.xlsx", sheet = "Complete transcriptome dataset", skip = 2)
```

```{r}
VBgam_vs_asexual <- readxl::read_excel("data/raw/published/12864_2019_6322_MOESM2_ESM.xlsx", sheet = "Gametocyte vs asexual hyb.", skip = 3)
```


```{r}
VBpublished_gam_markers <- readxl::read_excel("data/raw/published/12864_2019_6322_MOESM2_ESM.xlsx", sheet = "Published gametocyte markers", skip = 1)
```

```{r}
VBreg_functional <- readxl::read_excel("data/raw/published/12864_2019_6322_MOESM3_ESM.xlsx", sheet = "Regulatory and functional ass.", skip = 3)
```

```{r}
VBreg_functional <- VBreg_functional %>% 
rename_with(., ~paste0('Asex_Epi_',.), starts_with("H")) %>% 
rename_with(., ~paste0('AP2_',.), starts_with("PF3D7"))%>% 
rename_with(., ~paste0(., ' motif'), matches(c("ATGTGTA", "AGACA")))%>% 
mutate(across(everything(), ~str_replace(.,'√', 'Yes'))) %>%
  rename('ID'=`Gene ID`)
```

```{r}
stocha_dr_gns_VB_cluster <- stocha_pk_gns %>% 
left_join(., (VBreg_functional %>% dplyr::select(ID, 'Cluster')) , by = 'ID') %>% 
column_to_rownames('ID')
```

```{r}
##BF2
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)|!is.na(Bgam_sex_PK))  %>% pull(ID))) %>%
dplyr::select(`PlasmoDB ID`, starts_with('Day')) %>% 
column_to_rownames("PlasmoDB ID") %>%
as.matrix() %>%
pheatmap(., 
         cluster_cols = F,
         show_rownames = F,
         annotation_row = stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)|!is.na(Bgam_sex_PK))  %>% column_to_rownames('ID') %>% mutate(annot= coalesce(Bgam_sex_stocha_VKCK,Bgam_sex_PK)) %>% dplyr::select(annot))
```


```{r}
stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)|!is.na(Bgam_sex_PK))
```

```{r}
##BF1
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(!is.na(Asex_gam_stocha_VKCK)|!is.na(Asex_gam_PK))  %>% pull(ID))) %>%
dplyr::select(`PlasmoDB ID`, starts_with('Day')) %>% 
column_to_rownames("PlasmoDB ID") %>%
as.matrix() %>%
pheatmap(., 
         cluster_cols = F,
         show_rownames = F,
         annotation_row = stocha_pk_gns %>% filter(!is.na(Asex_gam_stocha_VKCK)|!is.na(Asex_gam_PK))  %>% column_to_rownames('ID') %>% mutate(annot= coalesce(Asex_gam_stocha_VKCK,Asex_gam_PK)) %>% dplyr::select(annot))
```


```{r}
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(Bgam_sex_stocha_VKCK == 'Female')  %>% pull(ID))) %>% 
pivot_longer(cols = starts_with('Day'), names_to = 'Days') %>% 
mutate(across(c('Days'), ~factor(.,levels = VBgam_vs_asexual %>% select(starts_with('Day')) %>% colnames()))) %>%
# group_by(`PlasmoDB ID`) %>%
ggplot(., aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`)) + 
geom_point() + 
geom_line() + 
# stat_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "lm", se = FALSE) +
# geom_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "gam", se = FALSE) + 
theme(legend.position = "none") 
```

```{r}
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(Bgam_sex_stocha_VKCK == 'Male')  %>% pull(ID))) %>% 
pivot_longer(cols = starts_with('Day'), names_to = 'Days') %>% 
mutate(across(c('Days'), ~factor(.,levels = VBgam_vs_asexual %>% select(starts_with('Day')) %>% colnames()))) %>%
# group_by(`PlasmoDB ID`) %>%
ggplot(., aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`)) + 
geom_point() + 
geom_line() + 
# stat_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "lm", se = FALSE) +
# geom_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "gam", se = FALSE) + 
theme(legend.position = "none") 
```

```{r}
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(Asex_gam_stocha_VKCK == 'Gametocyte')  %>% pull(ID))) %>% 
pivot_longer(cols = starts_with('Day'), names_to = 'Days') %>% 
mutate(across(c('Days'), ~factor(.,levels = VBgam_vs_asexual %>% select(starts_with('Day')) %>% colnames()))) %>%
# group_by(`PlasmoDB ID`) %>%
ggplot(., aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`)) + 
geom_point() + 
geom_line() + 
# stat_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "lm", se = FALSE) +
# geom_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "gam", se = FALSE) + 
theme(legend.position = "none") 
```


```{r}
TraProt_bias_CDB <- VBreg_functional %>% 
dplyr::select('ID','Female transcripts', 'Male transcripts', 'Female proteins', 'Male Proteins') %>%
dplyr::filter(if_any(-c('ID'), ~ !is.na(.))) %>%
mutate('Lasonder transcript sex bias' = case_when(`Female transcripts` == 'Yes'  & `Male transcripts` == 'Yes' ~ 'Both',
                                        `Female transcripts` == 'Yes' & is.na(`Male transcripts`) ~ 'Female',
                                        is.na(`Female transcripts`) & `Male transcripts` == 'Yes' ~ 'Male',
                                        is.na(`Female transcripts`) & is.na(`Male transcripts`) ~ 'None'),
      'Tao protein sex bias' = case_when(`Female proteins` == 'Yes'  & `Male Proteins` == 'Yes' ~ 'Both',
                                        `Female proteins` == 'Yes' & is.na(`Male Proteins`) ~ 'Female',
                                        is.na(`Female proteins`) & `Male Proteins` == 'Yes' ~ 'Male',
                                        is.na(`Female proteins`) & is.na(`Male Proteins`) ~ 'None')) %>%
right_join(., stocha_pk_gns , by='ID')
```

```{r}
TraProt_bias_CDB %>% filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(`Lasonder transcript sex bias`, Bgam_sex_stocha_VKCK) %>%
ggplot(., aes(x = `Lasonder transcript sex bias`, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```

```{r}
TraProt_bias_CDB %>% filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(`Tao protein sex bias`, Bgam_sex_stocha_VKCK) %>%
ggplot(., aes(x = `Tao protein sex bias`, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```


```{r}
stocha_dr_gns_VB <- stocha_pk_gns%>% 
left_join(., VBreg_functional %>%
          dplyr::select(ID,`AP2-G`, AP2_PF3D7_0934400, AP2_PF3D7_0611200, AP2_PF3D7_1317200, AP2_PF3D7_0516800, `Male transcripts`, `Female transcripts`, `Male Proteins`, `Female proteins`, Cluster), 
          by=c('ID'='ID')) %>%
mutate('Lasonder transcript bias' = case_when(`Female transcripts` == 'Yes'  & `Male transcripts` == 'Yes' ~ 'Both',
                                        `Female transcripts` == 'Yes' & is.na(`Male transcripts`) ~ 'Female',
                                        is.na(`Female transcripts`) & `Male transcripts` == 'Yes' ~ 'Male',
                                        is.na(`Female transcripts`) & is.na(`Male transcripts`) ~ 'None'),
      'Tao protein bias' = case_when(`Female proteins` == 'Yes'  & `Male Proteins` == 'Yes' ~ 'Both',
                                        `Female proteins` == 'Yes' & is.na(`Male Proteins`) ~ 'Female',
                                        is.na(`Female proteins`) & `Male Proteins` == 'Yes' ~ 'Male',
                                        is.na(`Female proteins`) & is.na(`Male Proteins`) ~ 'None'))
```


```{r}
lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}  
```


```{r}
print(GGally::ggpairs(
    stocha_dr_gns_VB %>% dplyr::select(ID, Bgam_sex_stocha_VKCK, `Lasonder transcript bias`, `Tao protein bias`)%>%
mutate(across(c('Bgam_sex_stocha_VKCK'), ~replace_na(.,'None') )) %>%
    tibble::column_to_rownames("ID"),# %>%
    # lower = list(continuous = GGally::wrap(
    #     ggplot(data = data, mapping = mapping) +
    #     geom_point(size = 0.3, alpha = 0.3) +
    #     geom_abline(slope = 1, intercept = 0, color = "blue")
    # )),
    title = paste0("Comparison "), 
                          xlab = paste("comp")
)
      )
    
```



```{r}
##ALL_DAYS
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% 
           filter(!is.na(Bgam_sex_stocha_VKCK)), by='ID') %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```

```{r}
##ALL_DAYS
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% 
           filter(!is.na(Asex_gam_stocha_VKCK)), by='ID') %>%
dplyr::count(Asex_gam_stocha_VKCK, Cluster) %>%
ggplot(., aes(x = Cluster, y = n, fill = Asex_gam_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```


```{r}
stocha_dr_gns_VB %>% 
filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster,`Tao protein bias`,`Lasonder transcript bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single')) +
facet_grid(`Lasonder transcript bias`~`Tao protein bias`)
```

```{r}
stocha_dr_gns_VB %>% 
filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster,`Lasonder transcript bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = `Lasonder transcript bias`)) +
geom_col(position = position_dodge2(preserve = 'single')) +
facet_grid(.~Bgam_sex_stocha_VKCK)
```

```{r}
stocha_dr_gns_VB %>% 
filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster,`Lasonder transcript bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single')) +
facet_grid(.~`Lasonder transcript bias`)
```



```{r}
stocha_dr_gns_VB %>% 
filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```

```{r}
VBreg_functional %>% 
filter(Cluster == '8') %>%
left_join(., stocha_pk_gns %>% 
            filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster, `Male transcripts`) %>%
ggplot(., aes(x = `Male transcripts`, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```


```{r}
VBreg_functional %>% 
dplyr::select(ID, 'Male transcripts','Female transcripts') %>%
mutate(
    'Transcript sex bias' = case_when(
        `Male transcripts` == 'Yes' & is.na(`Female transcripts`) ~ 'Male',
        `Female transcripts` == 'Yes'& is.na(`Male transcripts`) ~ 'Female',
        `Female transcripts` == 'Yes' & `Male transcripts` == 'Yes'~ 'Both',
)) %>%
filter(if_any(-c('ID'), ~!is.na(.))) %>%
  left_join(., stocha_pk_gns %>% 
            filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::count(`Transcript sex bias`, Bgam_sex_stocha_VKCK) %>%
ggplot(., aes(x = `Transcript sex bias`, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))

```


```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID','ATGTGTA motif','AGACA motif', 'Bgam_sex_stocha_VKCK') %>%
dplyr::count(Bgam_sex_stocha_VKCK, `ATGTGTA motif`, `AGACA motif`) 
```


```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID','Other motifs', 'Bgam_sex_stocha_VKCK') %>%
dplyr::count(Bgam_sex_stocha_VKCK, `Other motifs`) 
```

```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID',Bgam_sex_stocha_VKCK, 'Stabilized transcripts') %>%
dplyr::count(Bgam_sex_stocha_VKCK, `Stabilized transcripts`) 
```

```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID',Bgam_sex_stocha_VKCK, starts_with('AP2')) %>%
mutate(across(starts_with('AP2'), ~as.numeric(.))) %>%
dplyr::filter(if_any(starts_with('AP2'), ~ !is.na(.))) 
```

```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Asex_gam_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID',Asex_gam_stocha_VKCK, starts_with('AP2')) %>%
mutate(across(starts_with('AP2'), ~as.numeric(.))) %>%
dplyr::filter(if_any(starts_with('AP2'), ~ !is.na(.))) 
```


```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID',Bgam_sex_stocha_VKCK, starts_with('As')) %>%
dplyr::filter(if_any(starts_with('As'), ~ !is.na(.))) 
# %>%
# dplyr::select(-c(`Gene ID`)) %>% 
# count(everything())
```

```{r}
VBreg_functional %>% 
right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
dplyr::select('ID',Bgam_sex_stocha_VKCK, starts_with('As')) %>%
dplyr::filter(if_any(starts_with('As'), ~ !is.na(.))) %>%
dplyr::count(Bgam_sex_stocha_VKCK, `Asex_Epi_H3K36me3`) %>%
ggplot(., aes(x = `Asex_Epi_H3K36me3`, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single'))
```

```{r}
stocha_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% 
filter(ID %in% c('PF3D7_0404100', 'PF3D7_0516800', 'PF3D7_1429200', 'PF3D7_1315800', 'PF3D7_0603600', 'PF3D7_1241400', 'PF3D7_0715100'))
```



```{r}
# Road2UMAP <- readRDS("data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.2.RDS")
```

```{r}
Road2UMAP@meta.data %>% mutate(across('orig.ident', factor)) %>% rownames_to_column('bc')%>% group_by(orig.ident) %>% slice_max(nFeature_RNA)
```

```{r}
Road2UMAP@meta.data <- Road2UMAP@meta.data %>% mutate(across(where(is.factor), as.character)) 
```

```{r}
Road2UMAP_raw <- DietSeurat(
  Road2UMAP,
  counts = TRUE,
  data = T,
  scale.data = FALSE,
  features = NULL,
  assays = NULL,
  dimreducs = c('pca', 'umap', 'umap_test'),
  graphs = NULL
)

```

```{r}
DimPlot(Road2UMAP, reduction = "umap", dims = c(2,3))
```

```{r}
FeaturePlot(Road2UMAP, features = c('PF3D7-0613800', 'PF3D7-0802100'), dims = c(2,3))
```

```{r}
# options(repr.plot.width=30, repr.plot.height=6)
DimPlot(Road2UMAP, reduction = "umap", split.by = 'Stage', dims = c(2,3))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
DimPlot(Road2UMAP, reduction = "umap", group.by = 'anno_300cells', dims = c(2,3))
```

```{r}
Idents(Road2UMAP) <- 'Stage'
fVSm <- FindMarkers(Road2UMAP, ident.1 = 'gametocyte (female)', ident.2 = 'gametocyte (male)', min.pct = 0.20, logfc.threshold = log(2))
```

```{r}
ggvenn(
  list(VB_early = VBreg_functional %>% filter(Cluster == '8' & `Male transcripts` == 'Yes') %>% pull(ID),
    male_dr = stocha_pk_gns %>% filter(Bgam_sex_stocha_VKCK == 'Male')  %>% pull(ID) ,
    male_upregulated = fVSm %>% filter(avg_log2FC<0) %>% rownames() %>% str_replace(., '-', '_')),
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
  stroke_size = 0.5, set_name_size = 4
  )
```
##############################5th may 2022
```{r}
fVSm
```

```{r, eval=FALSE}
system(
  paste0('rsync -av jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/raw/Pf3d7_Genes_Summary_190322.csv /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/raw/')
  )
```

```{r}
Pf3D7_genes_plasmoDB <- read_csv("data/raw/Pf3d7_Genes_Summary_190322.csv") %>% 
dplyr::rename("gene_id" = `Gene ID` ) %>%
mutate(across(gene_id, ~str_replace(.,"_", "-")))
```

```{r}
head(Pf3D7_genes_plasmoDB)
```

```{r}
fVSm %>% 
rownames_to_column("gene_id") %>% 
left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Product Description", "Computed GO Components", "Computed GO Functions", 
"Computed GO Processes", "Curated GO Components", "Curated GO Functions", "Curated GO Processes")], by = "gene_id")
```

```{r}
Idents(Road2UMAP) <- 'anno_300cells'
lfVSlm <- FindMarkers(Road2UMAP, ident.1 = 'Gametocytes (late female)', ident.2 = 'Gametocytes (late male)', min.pct = 0.30, logfc.threshold = log(2))
```



```{r}
dr_sex_col2 <- brewer.pal(3,"Set1")[1:2]
names(dr_sex_col2) <- c('Male', 'Female')
```

```{r}
de_sex_col <- brewer.pal(3,"Set1")
names(de_sex_col) <- c('Male', 'Female', 'Not DE')
```

```{r}
dr_sex_col <- brewer.pal(4,"Set2")
names(dr_sex_col) <- c('Male', 'Female', 'Female,Male', 'No bias')
```

```{r}
dr_stage_col <- brewer.pal(4,"Set3")
names(dr_stage_col) <- c('Asexual', 'Gametocyte', 'Asexual,Gametocyte', 'No bias')
dr_stage_col2 <- brewer.pal(3,"Set2")[1:2]
names(dr_stage_col2) <- c('Asexual', 'Gametocyte')
```



```{r}
##ALL DAYS Bgam
lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id") %>%
mutate(across(gene_id, ~str_replace(., '-', '_') )) %>%
right_join(., stocha_dr_gns_VB_cluster %>%
           filter(!is.na(Bgam_sex_stocha_VKCK) ) %>% 
           rownames_to_column('gene_id') , by='gene_id') %>% 
dplyr::select(gene_id,p_val,avg_log2FC,pct.1,pct.2,p_val_adj,SexDE,Bgam_sex_stocha_VKCK) %>%
dplyr::count(SexDE, Bgam_sex_stocha_VKCK) %>%
mutate_if(is.character, ~replace_na(.,'Not DE')) %>%
ggplot(., aes(x = SexDE, y = n, fill = Bgam_sex_stocha_VKCK)) +
# geom_col(position = position_dodge2(preserve = 'single')) +
theme(axis.text=element_text(size=26), 
      axis.title=element_text(size=30,face="bold"), 
      legend.text = element_text(size=30), 
      legend.title = element_text(size=30,face="bold")) + 
rotate_x_text() +
labs(x = "Sex DE", y = 'Counts') + 
guides(fill=guide_legend(title="Sex \ncorrelation"))+ 
geom_col(position = position_dodge2(width = 0.2)) +
scale_fill_manual(values = dr_stage_col)+
geom_label(aes(label = n), position = position_dodge2(width = 01),size =8, angle = 90,colour = "white", fontface = "bold",show.legend = FALSE)
```


```{r}
##BF1 
lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id") %>%
mutate(across(gene_id, ~str_replace(., '-', '_') )) %>%
right_join(., stocha_dr_gns_VB_cluster %>%
           filter(!is.na(Asex_gam_stocha_VKCK) ) %>% 
           rownames_to_column('gene_id') , by='gene_id') %>% 
dplyr::select(gene_id,p_val,avg_log2FC,pct.1,pct.2,p_val_adj,SexDE,Asex_gam_stocha_VKCK) %>%
dplyr::count(SexDE, Asex_gam_stocha_VKCK) %>%
mutate_if(is.character, ~replace_na(.,'Not DE')) %>%
ggplot(., aes(x = SexDE, y = n, fill = Asex_gam_stocha_VKCK)) +
# geom_col(position = position_dodge2(preserve = 'single')) +
theme(axis.text=element_text(size=26), 
      axis.title=element_text(size=30,face="bold"), 
      legend.text = element_text(size=30), 
      legend.title = element_text(size=30,face="bold")) + 
rotate_x_text() +
labs(x = "Sex DE", y = 'Counts') + 
guides(fill=guide_legend(title="Sex \ncorrelation"))+ 
geom_col(position = position_dodge2(width = 0.2)) +
scale_fill_manual(values = dr_stage_col2)+
geom_label(aes(label = n), position = position_dodge2(width = 01),size =8, angle = 90,colour = "white", fontface = "bold",show.legend = FALSE)
```

```{r}
# VBreg_functional %>% 
# right_join(., stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)) , by='ID') %>%
# dplyr::select('ID',Bgam_sex, starts_with('As')) %>%
# dplyr::filter(if_any(starts_with('As'), ~ !is.na(.))) %>%
# dplyr::count(Bgam_sex, `Asex_Epi_H3K36me3`) %>%
# ggplot(., aes(x = `Asex_Epi_H3K36me3`, y = n, fill = Bgam_sex)) +
# geom_col(position = position_dodge2(preserve = 'single'))+
# theme(axis.text=element_text(size=14), 
#       axis.title=element_text(size=17,face="bold"), 
#       legend.text = element_text(size=17), 
#       legend.title = element_text(size=17,face="bold"))
```

```{r}
log(1)
```

```{r}
top10 <- lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id") %>% 
left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Product Description", "Computed GO Components", "Computed GO Functions", 
"Computed GO Processes", "Curated GO Components", "Curated GO Functions", "Curated GO Processes")], by = "gene_id")%>%
group_by(SexDE) %>%
    top_n(n = 10, wt = avg_log2FC)
```

```{r}
PfPDB_genes <- Pf3D7_genes_plasmoDB %>% mutate(across(gene_id, ~str_replace(., '-', '_')))%>% 
mutate(Gene = case_when(`Gene Name or Symbol` == 'N/A' ~ 'Conserved',
                       T ~ as.character(`Gene Name or Symbol`)))   %>% 
dplyr::select(gene_id, Gene) 
```

```{r}
all.genes <- rownames(Road2UMAP)

Road2UMAP <- ScaleData(Road2UMAP, features = all.genes)
```

```{r}
gams <- subset(Road2UMAP, idents = 'no assignment', invert=T) 

```

```{r}
gams_cat_levels <- c("Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)")
asx_gams_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)", "late trophozoite", "early schizont")
female_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)")
male_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early male)", "Gametocytes (late male)")
                  # c("early trophozoite", "Gametocytes (committed)", "late ring", "Gametocytes (branching)", "Gametocytes (developing)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)")
```

```{r}
asx_gams <- subset(Road2UMAP, subset = Stage %in% c('late schizont', 'early ring'), invert=T)
asx_gams@meta.data <- asx_gams@meta.data %>% 
  mutate(across(anno_300cells, ~case_when(. == 'Gametocytes (comitted)' ~ 'Gametocytes (committed)',
                                          TRUE ~ as.character(.))))%>% 
  mutate(stage_bfs = case_when(anno_300cells == 'no assignment' ~ as.character(Stage),
                                                     TRUE ~ as.character(anno_300cells))) %>% 
  mutate(asx_gam_short_labs = str_to_sentence(str_remove_all(stage_bfs, 'Gametocytes \\(|\\)$'))) %>% 
  dplyr::select(anno_300cells, rank2, rank, rank3,stage_bfs,asx_gam_short_labs, pseudo_asex, pseudo_male, pseudo_female,seurat_clusters, RNA_snn_res.0.2)%>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(stage_bfs, ~factor(.,levels = asx_gams_cat_levels))
          ) 

```

```{r}
female_asx_gams <- subset(Road2UMAP, subset = Stage %in% c('late schizont', 'early ring', 'gametocyte (male)', 'late trophozoite', 'early schizont'), invert=T)
female_asx_gams@meta.data <- female_asx_gams@meta.data %>% 
  mutate(across(anno_300cells, ~case_when(. == 'Gametocytes (comitted)' ~ 'Gametocytes (committed)',
                                          TRUE ~ as.character(.))))%>% 
  mutate(stage_bfs = case_when(anno_300cells == 'no assignment' ~ as.character(Stage),
                                                     TRUE ~ as.character(anno_300cells))) %>% 
  mutate(asx_gam_short_labs = str_to_sentence(str_remove_all(stage_bfs, 'Gametocytes \\(|\\)$'))) %>% 
  dplyr::select(anno_300cells, rank2, rank, rank3,stage_bfs,asx_gam_short_labs)%>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(stage_bfs, ~factor(.,levels = female_cat_levels))
          ) 

```

```{r}
male_asx_gams <- subset(Road2UMAP, subset = Stage %in% c('late schizont', 'early ring', 'gametocyte (female)', 'late trophozoite', 'early schizont'), invert=T)
male_asx_gams@meta.data <- male_asx_gams@meta.data %>% 
  mutate(across(anno_300cells, ~case_when(. == 'Gametocytes (comitted)' ~ 'Gametocytes (committed)',
                                          TRUE ~ as.character(.))))%>% 
  mutate(stage_bfs = case_when(anno_300cells == 'no assignment' ~ as.character(Stage),
                                                     TRUE ~ as.character(anno_300cells))) %>% 
  mutate(asx_gam_short_labs = str_to_sentence(str_remove_all(stage_bfs, 'Gametocytes \\(|\\)$'))) %>% 
  dplyr::select(anno_300cells, rank2, rank, rank3,stage_bfs,asx_gam_short_labs)%>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(stage_bfs, ~factor(.,levels = male_cat_levels))
          ) 

```

```{r}
DimPlot(male_asx_gams, dims = c(2,3), reduction = "umap", group.by = 'stage_bfs')
```

```{r}
gams@meta.data <- gams@meta.data %>%
  mutate(across(anno_300cells, ~case_when(. =='Gametocytes (comitted)' ~ 'Gametocytes (committed)',
                                           TRUE ~ as.character(anno_300cells))))%>% 
  mutate(gam_short_labs = str_to_sentence(str_remove_all(anno_300cells, 'Gametocytes \\(|\\)$'))) %>% 
  dplyr::select(anno_300cells, gam_short_labs, rank2, rank, rank3)
```

```{r}
gams@meta.data
```

```{r}
# gams_cat_levels <- c("Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)")
# asx_gams_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)", "late trophozoite", "early schizont")
# female_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)")
# male_cat_levels <- c("late ring","early trophozoite","Gametocytes (committed)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early male)", "Gametocytes (late male)")
```

```{r}
gams_cat_levels_short <- c("Committed", "Developing", "Branching", "Early female", "Late female", "Early male", "Late male")
asx_gams_cat_levels_short <- c("Late ring","Early trophozoite","Committed", "Developing", "Branching", "Early female", "Late female", "Early male", "Late male", "Late trophozoite", "Early schizont")
```

```{r}
gam_cols = c(brewer.pal(8,"Blues")[5], 
               brewer.pal(8,"Blues")[6], 
               brewer.pal(8,"Blues")[7], 
               brewer.pal(3,"Oranges")[2:3][1], 
               brewer.pal(3,"Oranges")[2:3][2], 
               brewer.pal(3,"Greens")[2:3][1], 
               brewer.pal(3,"Greens")[2:3][2])

asx_gam_cols = c(brewer.pal(9,"PuBuGn")[2], 
               brewer.pal(9,"PuBuGn")[3],
               brewer.pal(8,"Blues")[5], 
               brewer.pal(8,"Blues")[6], 
               brewer.pal(8,"Blues")[7], 
               brewer.pal(3,"Oranges")[2:3][1], 
               brewer.pal(3,"Oranges")[2:3][2], 
               brewer.pal(3,"Greens")[2:3][1], 
               brewer.pal(3,"Greens")[2:3][2], 
               brewer.pal(9,"PuBuGn")[7],
               brewer.pal(9,"PuBuGn")[8])
```


```{r}
female_asx_gam_cols = c(brewer.pal(9,"PuBuGn")[2], 
               brewer.pal(9,"PuBuGn")[3],
               brewer.pal(8,"Blues")[5], 
               brewer.pal(8,"Blues")[6], 
               brewer.pal(8,"Blues")[7], 
               brewer.pal(3,"Oranges")[2:3][1], 
               brewer.pal(3,"Oranges")[2:3][2])

male_asx_gam_cols = c(brewer.pal(9,"PuBuGn")[2], 
               brewer.pal(9,"PuBuGn")[3],
               brewer.pal(8,"Blues")[5], 
               brewer.pal(8,"Blues")[6], 
               brewer.pal(8,"Blues")[7], 
               brewer.pal(3,"Greens")[2:3][1], 
               brewer.pal(3,"Greens")[2:3][2])
```

```{r}
gam_cols_umap <- gam_cols
asx_gam_cols_umap <-asx_gam_cols
```

```{r}
female_cols_umap <- female_asx_gam_cols
male_gam_cols_umap <-male_asx_gam_cols
```

```{r}
names(gam_cols_umap) <- gams_cat_levels_short
names(asx_gam_cols_umap) <-asx_gams_cat_levels_short
```

```{r}
names(female_cols_umap) <- c("Late ring","Early trophozoite","Committed", "Developing", "Branching", "Early female", "Late female")
names(male_gam_cols_umap) <-c("Late ring","Early trophozoite","Committed", "Developing", "Branching", "Early male", "Late male")
```

```{r}
gam_cols
```

```{r}
gam_cols_umap
```

```{r, fig.width= 17, fig.height=8}
# options(repr.plot.width=17, repr.plot.height=8)
DimPlot(gams[,gams[["umap"]]@cell.embeddings[,'UMAP_3'] > -1.30], dims = c(2,3), pt.size = 1.5, label = T, label.size = 9, group = 'gam_short_labs')+
scale_color_manual(values = gam_cols_umap, labels = gams_cat_levels)+
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=20,face="bold"), 
      legend.text = element_text(size=28), 
      legend.title = element_text(size=17,face="bold"))
# options(repr.plot.width=7, repr.plot.height=6)
```

```{r, fig.width= 17, fig.height=8}
# options(repr.plot.width=17, repr.plot.height=8)
DimPlot(asx_gams, dims = c(2,3), pt.size = 1.5, label = T, label.size = 9, group = 'asx_gam_short_labs')+
scale_color_manual(values = asx_gam_cols_umap, labels = asx_gams_cat_levels_short)+
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=20,face="bold"), 
      legend.text = element_text(size=28), 
      legend.title = element_text(size=17,face="bold"))
# options(repr.plot.width=7, repr.plot.height=6)
```



```{r}
gams@meta.data <- gams@meta.data  %>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(anno_300cells, ~factor(.,levels = gams_cat_levels))
          ) 

# asx_gams@meta.data <- asx_gams@meta.data %>% 
#     mutate(across(c(rank2, rank), ~as.numeric(.)),
#            across(stage_bfs, ~factor(.,levels = asx_gams_cat_levels))
#           ) 
```

```{r}
# ggplot2::ggsave(filename = "feature.png")
```

```{r}
# options(repr.plot.width=21, repr.plot.height=10)
# DoHeatmap(gams, features = top10$gene_id) + NoLegend()
# # options(repr.plot.width=7, repr.plot.height=7)
```


```{r}
gams_cat_levels
```

```{r}
gam_cols[1]
```

```{r}
gams@meta.data  %>% 
    dplyr::select(anno_300cells, rank2) %>%
    arrange(anno_300cells,rank2)
```

```{r}
sorted_gam_scale_data_ffirst_bcodes <- gams@meta.data  %>% 
    dplyr::select(anno_300cells, rank2) %>% 
    mutate(across(rank2, ~as.numeric(.)),
           phmap_colors = case_when(
            anno_300cells == 'Gametocytes (committed)' ~ gam_cols[1],
            anno_300cells == 'Gametocytes (developing)'~ gam_cols[2],
            anno_300cells == 'Gametocytes (branching)'~ gam_cols[3],
            anno_300cells == 'Gametocytes (early female)'~ gam_cols[4],
            anno_300cells == 'Gametocytes (late female)'~ gam_cols[5],
            anno_300cells == 'Gametocytes (early male)'~ gam_cols[6],
            anno_300cells == 'Gametocytes (late male)'~ gam_cols[7] ),
           across(phmap_colors, ~factor(.,levels = gam_cols)),
          ) %>%
    arrange(anno_300cells,rank2)
```

```{r}
sorted_asx_gam_scale_data_ffirst_bcodes <- asx_gams@meta.data  %>% 
    dplyr::select(stage_bfs, rank, rank2) %>% 
    mutate(across(rank, ~as.numeric(.)),
           phmap_colors = case_when(
            stage_bfs == 'late ring' ~ asx_gam_cols[1],
            stage_bfs == 'early trophozoite'~ asx_gam_cols[2],
            stage_bfs == 'Gametocytes (committed)' ~ asx_gam_cols[3],
            stage_bfs == 'Gametocytes (developing)'~ asx_gam_cols[4],
            stage_bfs == 'Gametocytes (branching)'~ asx_gam_cols[5],
            stage_bfs == 'Gametocytes (early female)'~ asx_gam_cols[6],
            stage_bfs == 'Gametocytes (late female)'~ asx_gam_cols[7],
            stage_bfs == 'Gametocytes (early male)'~ asx_gam_cols[8],
            stage_bfs == 'Gametocytes (late male)'~ asx_gam_cols[9],
            stage_bfs == 'late trophozoite'~ asx_gam_cols[10],
            stage_bfs == 'early schizont'~ asx_gam_cols[11] ),
           across(phmap_colors, ~factor(.,levels = asx_gam_cols)),
          ) %>%
    arrange(stage_bfs,rank2, rank)
```

```{r}
female_sorted_asx_gam_scale_data_ffirst_bcodes <- female_asx_gams@meta.data  %>% 
    dplyr::select(stage_bfs, rank, rank2) %>% 
    mutate(across(rank, ~as.numeric(.)),
           phmap_colors = case_when(
            stage_bfs == 'late ring' ~ female_asx_gam_cols[1],
            stage_bfs == 'early trophozoite'~ female_asx_gam_cols[2],
            stage_bfs == 'Gametocytes (committed)' ~ female_asx_gam_cols[3],
            stage_bfs == 'Gametocytes (developing)'~ female_asx_gam_cols[4],
            stage_bfs == 'Gametocytes (branching)'~ female_asx_gam_cols[5],
            stage_bfs == 'Gametocytes (early female)'~ female_asx_gam_cols[6],
            stage_bfs == 'Gametocytes (late female)'~ female_asx_gam_cols[7]#,
            # stage_bfs == 'Gametocytes (early male)'~ asx_gam_cols[8],
            # stage_bfs == 'Gametocytes (late male)'~ asx_gam_cols[9],
            # stage_bfs == 'late trophozoite'~ asx_gam_cols[10],
            # stage_bfs == 'early schizont'~ asx_gam_cols[11] 
            ),
           across(phmap_colors, ~factor(.,levels = female_asx_gam_cols)),
          ) %>%
    arrange(stage_bfs,rank2, rank)
```

```{r}
male_sorted_asx_gam_scale_data_ffirst_bcodes <- male_asx_gams@meta.data  %>% 
    dplyr::select(stage_bfs, rank, rank2) %>% 
    mutate(across(rank, ~as.numeric(.)),
           phmap_colors = case_when(
            stage_bfs == 'late ring' ~ male_asx_gam_cols[1],
            stage_bfs == 'early trophozoite'~ male_asx_gam_cols[2],
            stage_bfs == 'Gametocytes (committed)' ~ male_asx_gam_cols[3],
            stage_bfs == 'Gametocytes (developing)'~ male_asx_gam_cols[4],
            stage_bfs == 'Gametocytes (branching)'~ male_asx_gam_cols[5],
            # stage_bfs == 'Gametocytes (early female)'~ asx_gam_cols[6],
            # stage_bfs == 'Gametocytes (late female)'~ asx_gam_cols[7],
            stage_bfs == 'Gametocytes (early male)'~ male_asx_gam_cols[6],
            stage_bfs == 'Gametocytes (late male)'~ male_asx_gam_cols[7]#,
            # stage_bfs == 'late trophozoite'~ asx_gam_cols[10],
            # stage_bfs == 'early schizont'~ asx_gam_cols[11] 
            ),
           across(phmap_colors, ~factor(.,levels = male_asx_gam_cols)),
          ) %>%
    arrange(stage_bfs, rank2, rank)
```

```{r}
sorted_gam_scale_data <- gams@assays$RNA@scale.data %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, sorted_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```

```{r}
sorted_asx_gam_scale_data <- asx_gams@assays$RNA@scale.data %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, sorted_asx_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```

```{r}
female_sorted_asx_gam_scale_data <- female_asx_gams@assays$RNA@scale.data %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```

```{r}
male_sorted_asx_gam_scale_data <- male_asx_gams@assays$RNA@scale.data %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```


```{r}
gams_annot <- gams@meta.data %>% dplyr::select(anno_300cells)
asx_gams_annot <- asx_gams@meta.data %>% dplyr::select(stage_bfs)
male_asx_gams_annot <- male_asx_gams@meta.data %>% dplyr::select(stage_bfs)
female_asx_gams_annot <- female_asx_gams@meta.data %>% dplyr::select(stage_bfs)
```


```{r}
met.brewer(name="Archambault")
```

pheatmap of drivers in gam branch per dataset

```{r}
stocha_gns %>% filter(if_any(contains('drivers'), ~str_detect(.,'all_days'))) 
```

```{r}
sorted_gam_scale_data_ffirst_bcodes %>% dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>% deframe
```

```{r}
sorted_asx_gam_scale_data_ffirst_bcodes %>% dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>% deframe
```

```{r}
female_sorted_asx_gam_scale_data_ffirst_bcodes %>% dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>% deframe
```


```{r}
##ONLY ALL_DAYS
CDgam_annot <- stocha_pk_gns %>%
filter(!is.na(Bgam_sex_stocha_VKCK)) %>%
left_join(., lfVSlm %>% 
          rownames_to_column("gene_id")  %>%
          mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male'),
                across(gene_id, ~str_replace(., '-', '_') )),
          by=c('ID'='gene_id')
          ) %>% 
mutate(across(SexDE, ~factor(replace_na(.,'Not DE'), levels = c('Female', 'Male', 'Not DE')))) %>%
column_to_rownames('ID') %>%
dplyr::select(Bgam_sex_stocha_VKCK, SexDE) %>%
rename('Sex correlation' = Bgam_sex_stocha_VKCK)
```

```{r}
##ONLY ALL_DAYS
asx_gam_annot <- stocha_pk_gns%>% 
  filter(!is.na(Asex_gam_stocha_VKCK) & !is.na(Asex_gam_stocha_VKCK_anti)) %>%
    column_to_rownames('ID') %>% 
    dplyr::select(Asex_gam_stocha_VKCK, Asex_gam_stocha_VKCK_anti) %>%
    rename('Stage correlation' = Asex_gam_stocha_VKCK, 'Stage anti-correlation' = Asex_gam_stocha_VKCK_anti)
```

```{r}
stocha_dr_gns_VB %>%
filter(ID %in% rownames(CDgam_annot)) %>% 
filter(!is.na(Bgam_sex_stocha_VKCK))  %>%
dplyr::count(Bgam_sex_stocha_VKCK, Cluster,`Lasonder transcript bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex_stocha_VKCK)) +
geom_col(position = position_dodge2(preserve = 'single')) +
facet_grid(.~`Lasonder transcript bias`)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
sorted_gam_scale_data[stocha_pk_gns %>% 
                      filter(!is.na(Bgam_sex_stocha_VKCK)) %>% 
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = CDgam_annot,
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
         annotation_colors = list(Gametocytes = sorted_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```


```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                      # filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>% 
                      # filter(is.na(CDgam_sex) | CDgam_sex != 'Female,Male') %>%
                      # filter(is.na(Bgam_sex) | Bgam_sex != 'Female,Male') %>% 
                      filter(!is.na(Bgam_sex_stocha_VKCK)) %>% 
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = CDgam_annot,
        annotation_col = sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
male_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                      filter(!is.na(Bgam_sex_stocha_VKCK)) %>% 
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
        show_colnames = F,
         annotation_row = CDgam_annot,
        annotation_col = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
female_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                      filter(!is.na(Bgam_sex_stocha_VKCK)) %>% 
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
        show_colnames = F,
         annotation_row = CDgam_annot,
        annotation_col = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
##Including asexual stages
sorted_asx_gam_scale_data[stocha_pk_gns%>% 
                            filter(!is.na(Asex_gam_stocha_VKCK) & !is.na(Asex_gam_stocha_VKCK_anti)) %>%
                            pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
        show_colnames = F,
         annotation_row = asx_gam_annot,
        annotation_col = sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stage' = stage_bfs) %>%dplyr::select(Stage),
         annotation_colors = list(Stage = sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe,
                                  `Stage correlation` = dr_stage_col,
                                  `Stage anti-correlation`=dr_stage_col),
        fontsize = 20)
         
```


```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data[CDgam_annot %>% 
                      filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>% rownames(),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = CDgam_annot %>% 
         filter(SexDE != 'Not DE' & `Sex correlation` == SexDE),
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
         annotation_colors = list(Gametocytes = sorted_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col2, 
                                  SexDE = de_sex_col[1:2]),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
##Bgam ALLDAYS
```





```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data[stocha_pk_gns %>% 
                        filter(!is.na(Bgam_sex_all_days) | !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
         annotation_colors = list(Gametocytes = sorted_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col,
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```




```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
male_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                        filter(!is.na(Bgam_sex_all_days) | !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
female_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                        filter(!is.na(Bgam_sex_all_days) | !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
male_sorted_asx_gam_scale_data[stocha_gns %>% 
                        filter(!is.na(Bgam_sex_all_days) & !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20#,
        # gaps_col = c(0,0,0,0,0,5,0)
        )
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
female_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                        filter(!is.na(Bgam_sex_all_days) & !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```



```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
male_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                        filter(!is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = male_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=10}
# options(repr.plot.width=12, repr.plot.height=5)
female_sorted_asx_gam_scale_data[stocha_pk_gns %>% 
                      # filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>% 
                      # filter(is.na(CDgam_sex) | CDgam_sex != 'Female,Male') %>%
                      # filter(is.na(Bgam_sex) | Bgam_sex != 'Female,Male') %>% 
                      # filter(if_any(matches('dir_all_d.*ale_in_gam'), ~.== 'correlated') ) %>% 
                        filter(!is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot,
        annotation_col = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rename('Stages' = stage_bfs) %>%dplyr::select(Stages),
         annotation_colors = list(Stages = female_sorted_asx_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(stage_bfs, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col, 
                                  `Sex anti-correlation` = dr_sex_col, 
                                  SexDE = de_sex_col),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```




```{r}
# Bgam_annot %>% 
# filter(SexDE != 'Not DE' & `Sex correlation` == SexDE)
```


```{r}
# sorted_gam_scale_data <- gams@assays$RNA@scale.data %>%
# as.data.frame() %>%
# rownames_to_column('barcode') %>%
# mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
# dplyr::select(barcode, sorted_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
# column_to_rownames('barcode') %>%
# as.matrix()
```

```{r}
pseudo_sorted_asx_gam_ffirst_bcodes <- asx_gams@meta.data  %>%
  dplyr::select(stage_bfs, rank, rank2,pseudo_asex, pseudo_male, pseudo_female) %>% 
  mutate(pseudo_mfa = coalesce(pseudo_male, pseudo_female, pseudo_asex),
         pseudo_am =coalesce(pseudo_asex, pseudo_male, pseudo_female),
         pseudo_af = coalesce(pseudo_asex, pseudo_male, pseudo_female)) %>%
  mutate(across(rank, ~as.numeric(.)),
           phmap_colors = case_when(
            stage_bfs == 'late ring' ~ asx_gam_cols[1],
            stage_bfs == 'early trophozoite'~ asx_gam_cols[2],
            stage_bfs == 'Gametocytes (committed)' ~ asx_gam_cols[3],
            stage_bfs == 'Gametocytes (developing)'~ asx_gam_cols[4],
            stage_bfs == 'Gametocytes (branching)'~ asx_gam_cols[5],
            stage_bfs == 'Gametocytes (early female)'~ asx_gam_cols[6],
            stage_bfs == 'Gametocytes (late female)'~ asx_gam_cols[7],
            stage_bfs == 'Gametocytes (early male)'~ asx_gam_cols[8],
            stage_bfs == 'Gametocytes (late male)'~ asx_gam_cols[9],
            stage_bfs == 'late trophozoite'~ asx_gam_cols[10],
            stage_bfs == 'early schizont'~ asx_gam_cols[11] ),
           across(phmap_colors, ~factor(.,levels = asx_gam_cols)),
          ) %>%
  arrange(stage_bfs,pseudo_asex, pseudo_male, pseudo_female)
```


```{r}
counts_asx_gam <- asx_gams@assays$RNA@counts %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, pseudo_sorted_asx_gam_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```

```{r}
DimPlot(asx_gams, reduction = "umap", dims = c(2,3), group.by = 'stage_bfs', label = T)
FeaturePlot(Road2UMAP, features = "pseudo_female", dims = c(2,3))
FeaturePlot(asx_gams, features = "pseudo_male", dims = c(2,3))
```


```{r}

Idents(asx_gams) <- 'stage_bfs'
asx_gams_bf1 <- subset(asx_gams, idents = c("late ring","early trophozoite","late trophozoite", "Gametocytes (committed)", "Gametocytes (developing)"))
gams_bf2 <- subset(asx_gams, idents = c("Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)"))

asx_gams_bf1@meta.data <- asx_gams_bf1@meta.data %>% mutate(across(stage_bfs, droplevels))
gams_bf2@meta.data <- gams_bf2@meta.data %>% mutate(across(stage_bfs, droplevels))


DimPlot(asx_gams_bf1, reduction = "umap", dims = c(2,3), group.by = 'stage_bfs', label = T)
DimPlot(gams_bf2, reduction = "umap", dims = c(2,3), group.by = 'stage_bfs', label = T)

asx_gams_bf1_sce <- SingleCellExperiment(assays = list(counts = as.matrix(asx_gams_bf1@assays$RNA@counts)), 
                                    colData = asx_gams_bf1@meta.data)
gams_bf2_sce <- SingleCellExperiment(assays = list(counts = as.matrix(gams_bf2@assays$RNA@counts)), 
                                    colData = gams_bf2@meta.data)

logcounts(asx_gams_bf1_sce) <- log2(counts(asx_gams_bf1_sce) + 1)
logcounts(gams_bf2_sce) <- log2(counts(gams_bf2_sce) + 1)
normcounts(asx_gams_bf1_sce) <- as.matrix(asx_gams_bf1@assays$RNA@data)
normcounts(gams_bf2_sce) <- as.matrix(gams_bf2@assays$RNA@data)
  
rowData(asx_gams_bf1_sce)$feature_symbol <- str_replace(rownames(asx_gams_bf1_sce), "-", "_")
rowData(gams_bf2_sce)$feature_symbol <- str_replace(rownames(gams_bf2_sce), "-", "_")

reducedDims(asx_gams_bf1_sce) <- SimpleList(PCA = asx_gams_bf1@reductions[["pca"]]@cell.embeddings, 
                                       UMAP = asx_gams_bf1@reductions[["umap"]]@cell.embeddings)
reducedDims(gams_bf2_sce) <- SimpleList(PCA = gams_bf2@reductions[["pca"]]@cell.embeddings, 
                                       UMAP = gams_bf2@reductions[["umap"]]@cell.embeddings)


asx_gams.sce <- SingleCellExperiment(assays = list(counts = as.matrix(asx_gams@assays$RNA@counts)), 
                                    colData = asx_gams@meta.data)

logcounts(asx_gams.sce) = log2(counts(asx_gams.sce) + 1)
normcounts(asx_gams.sce) <- as.matrix(asx_gams@assays$RNA@data)
# use gene names as feature symbols
rowData(asx_gams.sce)$feature_symbol <- str_replace(rownames(asx_gams.sce), "-", ".")

# asx_gams.sce <- selectFeatures(asx_gams.sce, suppress_plot = T, n_features = 250)
# asx_gams.sce = indexCell(asx_gams.sce)

reducedDims(asx_gams.sce) <- SimpleList(PCA = asx_gams@reductions[["pca"]]@cell.embeddings, 
                                       UMAP = asx_gams@reductions[["umap"]]@cell.embeddings)
```

```{r}
##Run slingshot to determine trajectories
asx_gams_bf1_sce<- slingshot(asx_gams_bf1_sce, clusterLabels = "stage_bfs", reducedDim = 'UMAP', start.clus = c('late ring'),
                         end.clus = c('Gametocytes (developing)', 'late trophozoite'), shrink =1, stretch = 0, allow.break = F
)
gams_bf2_sce<- slingshot(gams_bf2_sce, clusterLabels = "stage_bfs", reducedDim = 'UMAP', start.clus = c('Gametocytes (developing)'),
                         end.clus = c('Gametocytes (late male)', 'Gametocytes (late female)'), shrink =1, stretch = 0, allow.break = F
)

asx_gams.sce<- slingshot(asx_gams.sce, clusterLabels = "RNA_snn_res.0.2", reducedDim = 'UMAP', start.clus = c('1'),
                         end.clus = c("2", "8", '5'), shrink =1, stretch = 0, allow.break = F
)
```

```{r}
# colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
# plotcol <- colors[cut(asx_gams.sce$slingPseudotime_3, breaks=100)]

colData(asx_gams_bf1_sce)[,c('slingPseudotime_1', 'slingPseudotime_2')] %>% as.data.frame() %>% rownames_to_column('bcode') %>% write_csv(., 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_pts.csv')
colData(gams_bf2_sce)[,c('slingPseudotime_1', 'slingPseudotime_2')] %>% as.data.frame() %>% rownames_to_column('bcode') %>% write_csv(., 'data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_pts.csv')

asx_gams_bf1_sds <-SlingshotDataSet(asx_gams_bf1_sce)
gams_bf2_sds <- SlingshotDataSet(gams_bf2_sce)

plot(reducedDims(asx_gams_bf1_sce)$UMAP[,2:3], col = brewer.pal(5,"Set1")[as.factor(asx_gams_bf1_sce$stage_bfs)], pch=16, asp = 1)
lines(asx_gams_bf1_sds, lwd=2, col='black',dims = 2:3)

plot(reducedDims(gams_bf2_sce)$UMAP[,2:3], col = brewer.pal(6,"Set1")[as.factor(gams_bf2_sce$stage_bfs)], pch=16, asp = 1)
lines(gams_bf2_sds, lwd=2, col='black',dims = 2:3)

plot(reducedDims(asx_gams.sce)$UMAP[,2:3], col = brewer.pal(7,"Set1")[as.factor(asx_gams.sce$RNA_snn_res.0.2)], pch=16, asp = 1)
lines(SlingshotDataSet(asx_gams.sce), lwd=2, col='black')
# dev.off()
```

```{r}
asx_gams_bf1_ptime <- slingPseudotime(asx_gams_bf1_sds, na = FALSE)
gams_bf2_ptime <- slingPseudotime(gams_bf2_sds, na = FALSE)
asx_gams_bf1_cellWeights <- slingCurveWeights(asx_gams_bf1_sds)
gams_bf2_cellWeights <- slingCurveWeights(gams_bf2_sds)
```

```{r}
##save RDS for running in farm
saveRDS(asx_gams_bf1_sce, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_sce.RDS')
saveRDS(gams_bf2_sce, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_sce.RDS')

saveRDS(asx_gams_bf1_sds, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_sds.RDS')
saveRDS(gams_bf2_sds, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_sds.RDS')

saveRDS(asx_gams_bf1_ptime, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_ptime.RDS')
saveRDS(gams_bf2_ptime, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_ptime.RDS')

saveRDS(asx_gams_bf1_cellWeights, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_cellWeights.RDS')
saveRDS(gams_bf2_cellWeights, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_cellWeights.RDS')
```

```{r, eval=False}

system('rsync -auv /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/RoadtoUMAP/pseudotime_tradeseq/* jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/')

```

```{r, eval=FALSE}
##Run in farm

library(tradeSeq)
library(RColorBrewer)
library(SingleCellExperiment)
library(slingshot)
library(BiocParallel)

asx_gams_bf1_sce <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_sce.RDS')
gams_bf2_sce <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_sce.RDS')

asx_gams_bf1_sds <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_sds.RDS')
gams_bf2_sds <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_sds.RDS')

asx_gams_bf1_ptime <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_ptime.RDS')
gams_bf2_ptime <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_ptime.RDS')

asx_gams_bf1_cellWeights <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_cellWeights.RDS')
gams_bf2_cellWeights <- readRDS('/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_cellWeights.RDS')

asx_gams_bf1_eK <- evaluateK(counts = counts(asx_gams_bf1_sce), sds = asx_gams_bf1_sds, k = 3:9, nGenes = 200, verbose = T, parallel = T)
gams_bf2_eK <- evaluateK(counts = counts(gams_bf2_sce), sds = gams_bf2_sds, k = 3:9, nGenes = 200, verbose = T, parallel = T)

asx_gams_bf1_ts_sce <- fitGAM(counts = counts(asx_gams_bf1_sce), pseudotime = asx_gams_bf1_ptime, cellWeights = asx_gams_bf1_cellWeights, nknots = 4, verbose = T, parallel = T)
gams_bf2_ts_sce <- fitGAM(counts = counts(gams_bf2_sce), pseudotime = gams_bf2_ptime, cellWeights = gams_bf2_cellWeights, nknots = 4, verbose = T, parallel = T)

saveRDS(asx_gams_bf1_ts_sce, '/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_ts_sce.RDS')
saveRDS(gams_bf2_ts_sce, '/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_ts_sce.RDS')

saveRDS(asx_gams_bf1_ts_sce, '/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_ts_sce.RDS')
saveRDS(gams_bf2_ts_sce, '/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_ts_sce.RDS')
```

```{r, eval=False}
system('rsync -auv jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/gams_bf2_ts_sce.RDS /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/RoadtoUMAP/pseudotime_tradeseq/')
system('rsync -auv jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/asx_gams_bf1_ts_sce.RDS /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/RoadtoUMAP/pseudotime_tradeseq/')
```

```{r, eval=False}
gams_bf2_ts_sce <- readRDS('data/processed/RoadtoUMAP/pseudotime_tradeseq/gams_bf2_ts_sce.RDS')
asx_gams_bf1_ts_sce <- readRDS('data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_bf1_ts_sce.RDS')

```

```{r, eval=False}
plotGeneCount(curve = gams_bf2_sds, counts = counts(gams_bf2_sce), clusters = apply(slingClusterLabels(gams_bf2_sds), 1, which.max), models = gams_bf2_ts_sce)
```

```{r}
BF2Res <- earlyDETest(gams_bf2_ts_sce, knots = c(2, 1), l2fc=log2(2))
oBF2 <- order(BF2Res$waldStat, decreasing = TRUE)
head(rownames(BF2Res)[oBF2])
```


```{r}
plotSmoothers(gams_bf2_ts_sce, counts(gams_bf2_sce), gene = rownames(BF2Res)[oBF2][1])
```

```{r}
ggvenn(
  list(BF2_DE = BF2Res %>% arrange(desc(waldStat)) %>% head(n=500) %>% rownames() %>%str_replace(.,'-','_'),
    # male_dr_gambranch = stocha_pk_gns %>% filter(if_any(matches('dir.*_male_in_gambranch'), ~.== 'correlated') )  %>% pull(ID) ,
    male_rv = stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK)&!is.na(Bgam_sex_stocha_VKCK_anti))%>% pull(ID) ,
    male_ptime = stocha_pk_gns %>% filter(!is.na(Bgam_sex_PK)&!is.na(Bgam_sex_PK_anti)) %>% pull(ID)),
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
  stroke_size = 0.5, set_name_size = 4
  )
```


```{r}
female_up_male_down <- stocha_pk_gns  %>% 
  filter(Bgam_sex_stocha_VKCK == 'Female' & Bgam_sex_PK =='Female' & Bgam_sex_stocha_VKCK_anti =='Male' & Bgam_sex_PK_anti == 'Male')  %>%
  pull(ID) %>% str_replace(.,'_','-')
```

```{r}
male_up_female_down <- stocha_pk_gns  %>% 
  filter(Bgam_sex_stocha_VKCK == 'Male' & Bgam_sex_PK =='Male' & Bgam_sex_stocha_VKCK_anti =='Female' & Bgam_sex_PK_anti == 'Female')  %>%
  pull(ID)%>% str_replace(.,'_','-')
```

```{r}
plotSmoothers(gams_bf2_ts_sce, counts(gams_bf2_sce), gene = female_up_male_down[1]) + scale_fill_discrete(name='lineage', labels = c("A", "B"))
```

```{r, fig.width=30, fig.height=15}
female_up_male_down %>% 
map(. %>% plotSmoothers(gams_bf2_ts_sce, counts(gams_bf2_sce), gene = .) ) %>% 
plot_grid(plotlist = ., labels = female_up_male_down, ncol = 5) 
```


```{r, fig.width=30, fig.height=15}
male_up_female_down[1:20] %>% 
map(. %>% plotSmoothers(gams_bf2_ts_sce, counts(gams_bf2_sce), gene = .) ) %>% 
plot_grid(plotlist = ., labels = male_up_female_down[1:20], ncol = 5) 
```



```{r}
library(clusterExperiment)
nPointsClus <- 8
clusPat <- clusterExpressionPatterns(gams_bf2_ts_sce, nPoints = nPointsClus,
                                     genes = c(male_up_female_down, female_up_male_down))
clusterLabels <- primaryCluster(clusPat$rsec)
```

```{r}
male_up <- stocha_pk_gns  %>% 
  filter(Bgam_sex_stocha_VKCK == 'Male' | Bgam_sex_PK =='Male' )  %>%
  pull(ID)%>% str_replace(.,'_','-')
```

```{r}
nPointsClus <- 50
clusPat <- clusterExpressionPatterns(gams_bf2_ts_sce, nPoints = nPointsClus,
                                     genes = male_up)
clusterLabels <- primaryCluster(clusPat$rsec)
```

```{r}
cUniq <- unique(clusterLabels)
# cUniq <- cUniq[!cUniq == -1] # remove unclustered genes

for (xx in cUniq) {
  cId <- which(clusterLabels == xx)
  p <- ggplot(data = data.frame(x = 1:nPointsClus,
                                y = rep(range(clusPat$yhatScaled[cId, ]),
                                        nPointsClus / 2)),
              aes(x = x, y = y)) +
    geom_point(alpha = 0) +
    labs(title = paste0("Cluster ", xx),  x = "Pseudotime", y = "Normalized expression") +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5))
  for (ii in 1:length(cId)) {
    geneId <- rownames(clusPat$yhatScaled)[cId[ii]]
    p <- p +
      geom_line(data = data.frame(x = rep(1:nPointsClus, 2),
                                  y = clusPat$yhatScaled[geneId, ],
                                  lineage = rep(0:1, each = nPointsClus)),
                aes(col = as.character(lineage), group = lineage), lwd = 1.5)
  }
  p <- p + guides(color = FALSE) +
    scale_color_manual(values = c("orange", "darkseagreen3"),
                       breaks = c("0", "1"))  
  print(p)
}
```

```{r}
female_up <- stocha_pk_gns  %>% 
  filter(Bgam_sex_stocha_VKCK == 'Female' | Bgam_sex_PK =='Female' )  %>%
  pull(ID)%>% str_replace(.,'_','-')
```

```{r}
nPointsClus <- 10
clusPat <- clusterExpressionPatterns(gams_bf2_ts_sce, nPoints = nPointsClus,
                                     genes = c(female_up, male_up), ncores=6, nReducedDims=15, minSizes=3)
clusterLabels <- primaryCluster(clusPat$rsec)
```

```{r}
cUniq <- unique(clusterLabels)
# cUniq <- cUniq[!cUniq == -1] # remove unclustered genes

for (xx in cUniq) {
  cId <- which(clusterLabels == xx)
  p <- ggplot(data = data.frame(x = 1:nPointsClus,
                                y = rep(range(clusPat$yhatScaled[cId, ]),
                                        nPointsClus / 2)),
              aes(x = x, y = y)) +
    geom_point(alpha = 0) +
    labs(title = paste0("Cluster ", xx),  x = "Pseudotime", y = "Normalized expression") +
    theme_classic() +
    theme(plot.title = element_text(hjust = 0.5))
  for (ii in 1:length(cId)) {
    geneId <- rownames(clusPat$yhatScaled)[cId[ii]]
    p <- p +
      geom_line(data = data.frame(x = rep(1:nPointsClus, 2),
                                  y = clusPat$yhatScaled[geneId, ],
                                  lineage = rep(0:1, each = nPointsClus)),
                aes(col = as.character(lineage), group = lineage), lwd = 1.5)
  }
  p <- p + guides(color = FALSE) +
    scale_color_manual(values = c("orange", "darkseagreen3"),
                       breaks = c("0", "1"))  
  print(p)
}
```


```{r}
asx_gam_lin <- getLineages(asx_gams.sce, "RNA_snn_res.0.2",  reducedDim = 'UMAP', start.clus = c('1'), end.clus = c("2", "8", '5'))

# plot(reducedDims(asx_gams.sce)$UMAP, col = plotcol, pch=16, asp = 1)
plot(reducedDims(asx_gams.sce)$UMAP, col = brewer.pal(7,"Set1")[as.factor(asx_gams.sce$RNA_snn_res.0.2)], asp = 1, pch = 16)
# lines(SlingshotDataSet(asx_gams.sce), lwd=2, col='black')
# plot(asx_gams.sce, col = brewer.pal(9,"Set1")[cl], asp = 1, pch = 16)
lines(SlingshotDataSet(asx_gam_lin), lwd = 3, col = 'black', show.constraints = TRUE)
```

```{r}
asx_gam_curves <- getCurves(asx_gam_lin)

# plot(reducedDims(asx_gams.sce)$UMAP, col = plotcol, pch=16, asp = 1)
plot(reducedDims(asx_gams.sce)$UMAP, col = brewer.pal(7,"Set1")[as.factor(asx_gams.sce$RNA_snn_res.0.2)], asp = 1, pch = 16)
# lines(SlingshotDataSet(asx_gams.sce), lwd=2, col='black')
# plot(asx_gams.sce, col = brewer.pal(9,"Set1")[cl], asp = 1, pch = 16)
lines(SlingshotDataSet(asx_gam_curves), lwd = 3, col = 'black', show.constraints = TRUE)
```

```{r}
##save RDS for runningin farm
saveRDS(asx_gams.sce, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams.sce')
saveRDS(asx_gams_sds, 'data/processed/RoadtoUMAP/pseudotime_tradeseq/asx_gams_sds')
icMat <- evaluateK(counts = counts(asx_gams.sce), sds = asx_gams_sds, k = 3:10,nGenes = 200, verbose = T, parallel = T)
```

```{r, eval=False}
system('rsync -auv --partial-dir=.rsync-partial /Volumes/GoogleDrive/My\\ Drive/PHD_documents/malaria_phd/sc_velo/data/processed/RoadtoUMAP/pseudotime_tradeseq/* jr35@farm5-head1:/lustre/scratch118/malaria/team222/jr35/phd/V3_4TP/sc_velo/data/processed/RoadtoUMAP/ptime_tseq/')
```


```{r}
icMat <- evaluateK(counts = counts, sds = crv, k = 3:10, nGenes = 200, verbose = T)
BiocParallel::bpparam()
```
          
                   
```{r}
counts_asx_gam[stocha_pk_gns %>% 
                      # filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>% 
                      # filter(is.na(CDgam_sex) | CDgam_sex != 'Female,Male') %>%
                      # filter(is.na(Bgam_sex) | Bgam_sex != 'Female,Male') %>% 
                      # filter(if_any(matches('dir_all_d.*ale_in_gam'), ~.== 'correlated') ) %>% 
                        filter(!is.na(Bgam_sex_all_days) & !is.na(Bgam_sex_all_days_anti)) %>%
                      pull(ID),] %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column('bcode') %>%
  left_join(., pseudo_sorted_asx_gam_ffirst_bcodes %>% rownames_to_column('bcode') , by = 'bcode') %>%
    mutate(Male_lineage = case_when(!stage_bfs %in% c('Gametocytes (early female)', 'Gametocytes (late female)') ~ as.character(stage_bfs)),
           Female_lineage = case_when(!stage_bfs %in% c('Gametocytes (early male)', 'Gametocytes (late male)') ~  as.character(stage_bfs)),
           coarse = case_when(stage_bfs %in% c('Gametocytes (early female)', 'Gametocytes (late female)') ~ 'Female',
                              stage_bfs %in% c('Gametocytes (early male)', 'Gametocytes (late male)') ~ 'Male',
                              stage_bfs %in% c( 'Gametocytes (developing)', 'Gametocytes (committed)', 'Gametocytes (branching)') ~ 'Sex indeterminate',
                              stage_bfs %in% c('late ring', 'early trophozoite','late trophozoite', 'early schizont') ~ 'Asex'))%>% 
  # pivot_longer(.,cols = c(Male_lineage, Female_lineage), names_to = 'Lineage', values_to = 'stage_lineage')%>%
  # pivot_longer(.,cols = c(pseudo_am, pseudo_af), names_to = 'pseudo_sex', values_to = 'pseudo_val') %>%
  as.data.frame() %>%
  ggplot(., aes(x = pseudo_mfa,  y = log(PF3D7_1465900+1), color = coarse)) + 
  geom_point() +
  geom_smooth()
```

                      
```{r}
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(if_any(matches('dir_all_d.*_female_in_gambranch'), ~.== 'correlated') ) %>% pull(ID))) %>% 
pivot_longer(cols = starts_with('Day'), names_to = 'Days') %>% 
mutate(across(c('Days'), ~factor(.,levels = VBgam_vs_asexual %>% select(starts_with('Day')) %>% colnames()))) %>%
# group_by(`PlasmoDB ID`) %>%
ggplot(., aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`)) + 
geom_point() + 
geom_line() + 
# stat_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "lm", se = FALSE) +
# geom_smooth(aes(x = Days,  y = value, group = `PlasmoDB ID`, color = `PlasmoDB ID`), method = "gam", se = FALSE) + 
theme(legend.position = "none") 
```

```{r}
# female_sorted_asx_gam_scale_data <- female_asx_gams@assays$RNA@scale.data %>%
# as.data.frame() %>%
# rownames_to_column('barcode') %>%
# mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
# dplyr::select(barcode, female_sorted_asx_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
# column_to_rownames('barcode') %>%
# as.matrix()
```

```{r}
# male_sorted_asx_gam_scale_data <- male_asx_gams@assays$RNA@scale.data %>%
# as.data.frame() %>%
# rownames_to_column('barcode') %>%
# mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
# dplyr::select(barcode, male_sorted_asx_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
# column_to_rownames('barcode') %>%
# as.matrix()
```

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data[Bgam_annot %>% 
                      filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>% rownames(),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = Bgam_annot %>% 
         filter(SexDE != 'Not DE' & `Sex correlation` == SexDE),
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
         annotation_colors = list(Gametocytes = sorted_gam_scale_data_ffirst_bcodes %>% 
                                  dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%
                                  deframe, 
                                  `Sex correlation` = dr_sex_col2, 
                                  SexDE = de_sex_col[1:2]),
        fontsize = 20)
         
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
sorted_gam_scale_data_ffirst_bcodes2 <- gams@meta.data %>% 
dplyr::select(anno_300cells, rank2) %>% 
mutate(across(rank2, ~as.numeric(.)),
       across(anno_300cells, ~factor(.,levels = c("Gametocytes (comitted)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)")))) %>%
arrange(anno_300cells,rank2) 
```

```{r}
# sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes)
```

```{r}
sorted_gam_scale_data_ffirst2 <- gams@assays$RNA@scale.data %>%
as.data.frame() %>%
rownames_to_column('barcode') %>%
mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
dplyr::select(barcode, sorted_gam_scale_data_ffirst_bcodes %>% rownames()) %>%
column_to_rownames('barcode') %>%
as.matrix()
```

```{r, fig.width= 12, fig.height=7}
##Including all days pluy others
# options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data_ffirst2[stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% dplyr::select(CDgam_sex, `Gene ID`) %>% column_to_rownames('Gene ID'))%>% rename('Sex correlation' = CDgam_sex),
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        fontsize = 20)
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data_ffirst2[stocha_pk_gns %>% filter(if_any(contains('drivers'), ~str_detect(.,'all$'))) %>% filter(!is.na(Bgam_sex_stocha_VKCK)) %>% pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         annotation_row = (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% dplyr::select(CDgam_sex, `Gene ID`) %>% column_to_rownames('Gene ID'))%>% rename('Sex correlation' = CDgam_sex),
        annotation_col = sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        fontsize = 20)
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)
sorted_gam_scale_data[stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% pull(ID),] %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
        show_colnames = F,
         annotation_row = stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% column_to_rownames('Gene ID') %>% dplyr::select(Bgam_sex)%>% rename('Sex correlation' = Bgam_sex),
        annotation_col = gams@meta.data %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        fontsize = 20)
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
VBgam_vs_asexual %>% 
filter(`PlasmoDB ID` %in% (stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% pull(ID))) %>%
dplyr::select(`PlasmoDB ID`, starts_with('Day')) %>% 
column_to_rownames("PlasmoDB ID") %>%
as.matrix() %>%
pheatmap(., 
         cluster_cols = F, 
         show_rownames = F,
         annotation_row = stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% column_to_rownames('Gene ID') %>% dplyr::select(Bgam_sex))
```

```{r}
stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% arrange(CDgam_sex)
```

```{r}
Bgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE)
```

```{r}
# VBreg_functional %>% 
# right_join(., stocha_pk_gns %>%
#            filter(ID %in% (Bgam_annot %>% 
#                   filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
#                  rownames()))
```

```{r}
stocha_pk_gns %>%
filter(ID %in% (Bgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE)%>%
rownames()))
```

```{r, fig.width= 12, fig.height=7}
##CD ALL_DAYS ONLY
options(repr.plot.width=9, repr.plot.height=7)

stocha_dr_gns_VB%>%
filter(ID %in% (CDgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
rownames()))%>%
dplyr::count(CDgam_sex, Cluster,`Tao protein bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = CDgam_sex)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
facet_grid(.~`Tao protein bias`, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
theme(axis.text=element_text(size=17), 
      axis.title=element_text(size=27,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 22),
      plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
rotate_x_text() +
labs(x = "vanBiljon et al., clusters", y = 'Counts', title = 'Lasonder et al protein sex bias') + 
guides(fill=guide_legend(title="Sex \ncorrelation \nin committed\n& developing\ngametocytes"))+
scale_fill_manual(values = dr_sex_col2)

# options(repr.plot.width=7, repr.plot.height=7)
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
# VBreg_functional %>% 
# right_join(., stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') ) , by='ID') %>%
# count(Bgam_sex, Cluster) %>%
# ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex)) +
# geom_col(position = position_dodge2(preserve = 'single'))
```

```{r, fig.width= 9, fig.height=7}
##BGAM ALL_DAYS
options(repr.plot.width=9, repr.plot.height=7)

stocha_dr_gns_VB%>%
filter(ID %in% (Bgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
rownames()))%>%
dplyr::count(Bgam_sex, Cluster,`Tao protein bias`) %>%
ggplot(., aes(x = Cluster, y = n, fill = Bgam_sex)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
facet_grid(.~`Tao protein bias`, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
theme(axis.text=element_text(size=17), 
      axis.title=element_text(size=27,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 22),
      plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
rotate_x_text() +
labs(x = "vanBiljon et al., clusters", y = 'Counts', title = 'Lasonder et al protein sex bias') +
scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="Sex \ncorrelation \nin branching\ngametocytes"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
Pf3D7_genes_plasmoDB
```

```{r}
CDgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>% 
rownames_to_column('gene_id') %>% 
left_join(., Pf3D7_genes_plasmoDB %>% mutate(across(gene_id, ~str_replace(.,'-','_'))), by='gene_id') %>% 
dplyr::select(gene_id, 'Sex correlation','Gene Name or Symbol','# Exons in Gene','Transcript Length','Computed GO Functions','Curated GO Functions', )
```

```{r}
Bgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>% 
rownames_to_column('gene_id') %>% 
left_join(., Pf3D7_genes_plasmoDB %>% mutate(across(gene_id, ~str_replace(.,'-','_'))), by='gene_id') %>% 
dplyr::select(gene_id, 'Sex correlation','Gene Name or Symbol','# Exons in Gene','Transcript Length','Computed GO Functions','Curated GO Functions' )
```

```{r}
# options(repr.plot.width=7, repr.plot.height=7)
ggvenn(list('Committed & Developing \ngametocytes' = CDgam_annot %>% filter(SexDE != 'Not DE' & `Sex correlation` == SexDE)  %>% rownames(),
            'Branching \ngametocytes' = Bgam_annot %>% filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>% rownames() 
            ),
       fill_color = c("#0073C2FF", "#EFC000FF"),
       stroke_size = 0.7, set_name_size = 8, text_size = 8
      )+ 
scale_y_continuous(expand = expansion(mult = .3))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
###Morrilo hdp1 KO
```

```{r}
# hpd1KO <- readxl::read_excel("data/raw/published/morillo_hdp1KO/41564_2021_1045_MOESM2_ESM.xls", sheet =  "Data Set1 - Filtered DGE",  skip = 1)
```

```{r}
hdp1KO <- read_csv('data/raw/published/morillo_hdp1KO/hdp1KO.csv', skip = 1)
```

```{r}
hdp1KO %>% head
```

```{r}
stocha_pk_gns%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('ID'='geneID')
          )%>%
filter(ID %in% (Bgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
rownames()))%>%
dplyr::count(Bgam_sex, hdp1KO)
```

```{r}
stocha_dr_gns%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('ID'='geneID')
          )%>%
filter(ID %in% (CDgam_annot %>% 
filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
rownames()))%>%
dplyr::count(CDgam_sex, hdp1KO)%>%
ggplot(., aes(x = CDgam_sex, y = n, fill = hdp1KO)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 20)) + 
# rotate_x_text() +
# labs(x = "vanBiljon et al., cluster", y = 'Counts') +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="hdpKO"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
stocha_dr_gns%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('ID'='geneID')
          )%>%
filter(ID %in% (Bgam_annot %>%
                       # filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
                       rownames())
      )%>%
dplyr::count(Bgam_sex, hdp1KO)%>%
ggplot(., aes(x = Bgam_sex, y = n, fill = hdp1KO)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 20)) + 
# rotate_x_text() +
# labs(x = "vanBiljon et al., cluster", y = 'Counts') +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="hdpKO"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
stocha_dr_gns%>%
filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('ID'='geneID')
          )%>%
filter(!is.na(hdp1KO)) %>%
dplyr::count(CDgam_sex,Bgam_sex,hdp1KO)
```

```{r}
stocha_dr_gns%>%
filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('ID'='geneID')
          )%>%
filter(!is.na(hdp1KO))
```

```{r}
stocha_corr_genes %>%
map(. %>% filter(ID %in% c('PF3D7_0402000', 'MESA', 'PF3D7_0901700')))
```

```{r}
fVSm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id")  %>%
mutate(across(gene_id, ~str_replace(., '-', '_') )) %>%
right_join(.,stocha_dr_gns, by = c('gene_id'='ID'))%>%
filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('gene_id'='geneID')
          )%>%
filter(!is.na(hdp1KO)) %>%
dplyr::count(CDgam_sex,Bgam_sex,hdp1KO,SexDE)%>%
ggplot(., aes(x = hdp1KO, y = n, fill = SexDE)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =1.0, hjust =0.5, size =6)+
# facet_grid(.~CDgam_sex, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
facet_grid(CDgam_sex~Bgam_sex) +
theme(axis.text=element_text(size=18), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=20,face="bold"),
      strip.text.x = element_text(size = 20),
      strip.text.y = element_text(size = 20),
     plot.title = element_text(size = 20,hjust = 0.5,face="bold") ) + 
rotate_x_text() +
labs(x = "Morillo et al., HDP1 \nKO effect on gene", y = 'Counts',title = "Sex correlation in branching \ngametocytes") +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="SexDE"))+ 
scale_y_continuous(sec.axis = sec_axis(~ . , name = "Sex correlation in committed & \n developing gametocytes", breaks = NULL, labels = NULL))
```

```{r}
fVSm %>% arrange(desc(p_val_adj))
```

```{r}
lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id")  %>%
mutate(across(gene_id, ~str_replace(., '-', '_') ))%>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('gene_id'='geneID')
          )%>%
dplyr::count(SexDE,hdp1KO)%>%
ggplot(., aes(x = hdp1KO, y = n, fill = SexDE)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
# facet_grid(.~CDgam_sex, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
# facet_grid(.~CDgam_sex) +
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 20)) + 
rotate_x_text() +
# labs(x = "vanBiljon et al., cluster", y = 'Counts') +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="Sex DE"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id")  %>%
mutate(across(gene_id, ~str_replace(., '-', '_') )) %>%
right_join(.,stocha_dr_gns, by = c('gene_id'='ID'))%>%
filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>%
  filter(!is.na(CDgam_sex)) %>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('gene_id'='geneID')
          )%>%
filter(!is.na(hdp1KO)) %>%
dplyr::count(CDgam_sex,hdp1KO)%>%
ggplot(., aes(x = hdp1KO, y = n, fill = CDgam_sex)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
# facet_grid(.~CDgam_sex, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
# facet_grid(.~CDgam_sex) +
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 20)) + 
rotate_x_text() +
# labs(x = "vanBiljon et al., cluster", y = 'Counts') +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="Sex correlation"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
lfVSlm %>% mutate(SexDE = case_when(avg_log2FC > 0 ~ 'Female',
                                        avg_log2FC < 0 ~ 'Male')) %>%
rownames_to_column("gene_id")  %>%
mutate(across(gene_id, ~str_replace(., '-', '_') )) %>%
right_join(.,stocha_dr_gns, by = c('gene_id'='ID'))%>%
filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))%>%
  filter(!is.na(Bgam_sex)) %>%
left_join(., hdp1KO %>% 
          mutate(hdp1KO = case_when(`log2(∆hdp1/NF54) Fold-change` > 0 ~ 'Upregulated',
                                    `log2(∆hdp1/NF54) Fold-change` < 0 ~ 'Downregulated')
                 ),
          by=c('gene_id'='geneID')
          )%>%
filter(!is.na(hdp1KO)) %>%
dplyr::count(Bgam_sex,hdp1KO)%>%
ggplot(., aes(x = hdp1KO, y = n, fill = Bgam_sex)) +
geom_col(position = position_dodge2(preserve = 'single'))+ 
geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
# facet_grid(.~CDgam_sex, labeller = labeller(.cols  = c('Female' = 'Female protein', 'Male' = 'Male protein', 'None' = 'No bias'))) +
# facet_grid(.~CDgam_sex) +
theme(axis.text=element_text(size=16), 
      axis.title=element_text(size=22,face="bold"), 
      legend.text = element_text(size=22), 
      legend.title = element_text(size=22,face="bold"),
      strip.text.x = element_text(size = 20)) + 
rotate_x_text() +
# labs(x = "vanBiljon et al., cluster", y = 'Counts') +
# scale_fill_manual(values = dr_sex_col2)+ 
guides(fill=guide_legend(title="Sex correlation"))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
Josling_ap2g_Comparisons <- excel_sheets("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM9_ESM.xlsx") %>% 
  set_names() %>% 
  map_df(~read_xlsx("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM9_ESM.xlsx",.), .id = 'Study_comparison') %>% 
  select(where(~!all(is.na(.)))) %>% select(-c("P-value", "Total", "#AP2-G bound", "...8", "...10"))  %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% 
  dplyr::rename_all(~str_replace_all(.,"\r\n", ""))

```

```{r}
Josling_ap2g_sc_schizonts <- readxl::read_excel("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM6_ESM.xlsx", sheet = "Consensus peaks") %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% dplyr::rename_all(~str_replace_all(.,"\r\n", "")) %>%
  mutate(across(everything(), na_if, "NA"))
```

```{r}
Josling_ap2g_sc_ring <- readxl::read_excel("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM7_ESM.xlsx", sheet = "Consensus peaks") %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% dplyr::rename_all(~str_replace_all(.,"\r\n", ""))%>%
  mutate(across(everything(), na_if, "NA"))
```

```{r}
Josling_ap2g_stg1_gam <- readxl::read_excel("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM8_ESM.xlsx", sheet = "Consensus peaks") %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% dplyr::rename_all(~str_replace_all(.,"\r\n", ""))%>%
  mutate(across(everything(), na_if, "NA"))
```

```{r}
Josling_ap2g_ncc <- readxl::read_excel("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM10_ESM.xlsx", sheet = "1 NCC consensus peaks") %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% dplyr::rename_all(~str_replace_all(.,"\r\n", ""))%>%
  mutate(across(everything(), na_if, "NA"))
```

```{r}
Josling_ap2g_scc <- readxl::read_excel("data/raw/published/josling_ap2g_chip/41467_2020_15026_MOESM10_ESM.xlsx", sheet = "4 SCC consensus peaks") %>% dplyr::rename_all(~str_replace_all(.," ", "_")) %>% dplyr::rename_all(~str_replace_all(.,"\r\n", ""))%>%
  mutate(across(everything(), na_if, "NA"))
```

```{r}
stocha_dr_gns%>%
    left_join(., Josling_ap2g_PoranComp ,
              by=c('Gene ID'='Gene')
    )%>% 
    dplyr::count(lR_eT_Asex_gam, Upstream_peak_in_committed_schizonts)
```

```{r}
stocha_pk_gns %>% left_join(Josling_ap2g_stg1_gam %>% unite('id', c(left_id,right_id), sep = '_', na.rm = TRUE, remove = F) %>% pivot_longer(cols = c(left_id, right_id),names_to = 'relative_pos',values_to = 'ID', values_drop_na = T) , by = 'ID') %>% filter(!is.na(id)) %>%
    filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))  %>% dplyr::count(relative_pos, CDgam_sex, lR_eT_Asex_gam) %>% ggplot(., aes(x = lR_eT_Asex_gam, y = n, fill = CDgam_sex)) +
    geom_col(position = position_dodge2(preserve = 'single'))+ 
    geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
    facet_grid(.~relative_pos, labeller = labeller(.cols  = c('right_id' = 'right', 'left_id' = 'left'))) +
    theme(axis.text=element_text(size=17), 
          axis.title=element_text(size=27,face="bold"), 
          legend.text = element_text(size=22), 
          legend.title = element_text(size=22,face="bold"),
          strip.text.x = element_text(size = 22),
          plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
    rotate_x_text() +
    labs(x = "Sexual commitment \ncorrelation \nin early rings and late trophozoites", y = 'Counts', title = 'Relative position of gene to AP2G peak') + 
    guides(fill=guide_legend(title="Sex \ncorrelation \nin committed\n& developing\ngametocytes"))
```

```{r}
stocha_pk_gns %>% left_join(Josling_ap2g_stg1_gam %>% unite('id', c(left_id,right_id), sep = '_', na.rm = TRUE, remove = F) %>% pivot_longer(cols = c(left_id, right_id),names_to = 'relative_pos',values_to = 'ID', values_drop_na = T) , by = 'ID') %>% filter(!is.na(id)) %>%
    filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))  %>% dplyr::count(relative_pos, CDgam_sex, lR_eT_Asex_gam) %>% ggplot(., aes(x = lR_eT_Asex_gam, y = n, fill = CDgam_sex)) +
    geom_col(position = position_dodge2(preserve = 'single'))+ 
    geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
    facet_grid(.~relative_pos, labeller = labeller(.cols  = c('right_id' = 'right', 'left_id' = 'left'))) +
    theme(axis.text=element_text(size=17), 
          axis.title=element_text(size=27,face="bold"), 
          legend.text = element_text(size=22), 
          legend.title = element_text(size=22,face="bold"),
          strip.text.x = element_text(size = 22),
          plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
    rotate_x_text() +
    labs(x = "Sexual commitment \ncorrelation \nin early rings and late trophozoites", y = 'Counts', title = 'Relative position of gene to AP2G peak') + 
    guides(fill=guide_legend(title="Sex \ncorrelation \nin committed\n& developing\ngametocytes"))
```

```{r}
stocha_pk_gns %>% left_join(Josling_ap2g_sc_ring %>% unite('id', c(left_id,right_id), sep = '_', na.rm = TRUE, remove = F) %>% pivot_longer(cols = c(left_id, right_id),names_to = 'relative_pos',values_to = 'ID', values_drop_na = T) , by = 'ID') %>% filter(!is.na(id)) %>%
    filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))  %>% dplyr::count(relative_pos, CDgam_sex, lR_eT_Asex_gam) %>% ggplot(., aes(x = lR_eT_Asex_gam, y = n, fill = CDgam_sex)) +
    geom_col(position = position_dodge2(preserve = 'single'))+ 
    geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
    facet_grid(.~relative_pos, labeller = labeller(.cols  = c('right_id' = 'right', 'left_id' = 'left'))) +
    theme(axis.text=element_text(size=17), 
          axis.title=element_text(size=27,face="bold"), 
          legend.text = element_text(size=22), 
          legend.title = element_text(size=22,face="bold"),
          strip.text.x = element_text(size = 22),
          plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
    rotate_x_text() +
    labs(x = "Sexual commitment \ncorrelation \nin early rings and late trophozoites", y = 'Counts', title = 'Relative position of gene to AP2G peak') + 
    guides(fill=guide_legend(title="Sex \ncorrelation \nin committed\n& developing\ngametocytes"))
```

```{r}
stocha_pk_gns %>% left_join(Josling_ap2g_ncc %>% unite('id', c(left_id,right_id), sep = '_', na.rm = TRUE, remove = F) %>% pivot_longer(cols = c(left_id, right_id),names_to = 'relative_pos',values_to = 'ID', values_drop_na = T) , by = 'ID') %>% filter(!is.na(id)) %>%
    filter(if_any(contains('drivers'), ~str_detect(.,'all_days')))  %>% dplyr::count(relative_pos, CDgam_sex, lR_eT_Asex_gam) %>% ggplot(., aes(x = lR_eT_Asex_gam, y = n, fill = CDgam_sex)) +
    geom_col(position = position_dodge2(preserve = 'single'))+ 
    geom_text(aes(label = n), position = position_dodge2(width = 0.8), vjust =0.5, hjust =0.5, size =6)+
    facet_grid(.~relative_pos, labeller = labeller(.cols  = c('right_id' = 'right', 'left_id' = 'left'))) +
    theme(axis.text=element_text(size=17), 
          axis.title=element_text(size=27,face="bold"), 
          legend.text = element_text(size=22), 
          legend.title = element_text(size=22,face="bold"),
          strip.text.x = element_text(size = 22),
          plot.title = element_text(size = 27,hjust = 0.5,face="bold")) + 
    rotate_x_text() +
    labs(x = "Sexual commitment \ncorrelation \nin early rings and late trophozoites", y = 'Counts', title = 'Relative position of gene to AP2G peak') + 
    guides(fill=guide_legend(title="Sex \ncorrelation \nin committed\n& developing\ngametocytes"))
```

```{r, fig.width= 21, fig.height=10}
options(repr.plot.width=21, repr.plot.height=10)
DoHeatmap(gams, features = stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% arrange(CDgam_sex) %>% pull(ID) %>% str_replace(., '_', '-')) + NoLegend()
# options(repr.plot.width=7, repr.plot.height=7)
```

Replicate heatmaps for the different days to rule out batch effects as days are not intergrated

```{r}
Road2UMAP_list <- SplitObject(Road2UMAP, split.by = "orig.ident")
```

```{r}
# VlnPlot(D10.mrna, features = "PF3D7-0935390")
Road2UMAP_list_all_genes <- Road2UMAP_list %>% map(. %>% rownames())
```

```{r}
Road2UMAP_list <- Road2UMAP_list %>% map2(.,Road2UMAP_list_all_genes, ~ScaleData(.x, features = .y))
```

```{r}
gams_list <- map(Road2UMAP_list, ~subset(., idents = 'no assignment', invert=T))
```

```{r}
names(gams_list)
```

```{r, fig.width= 21, fig.height=14}
options(repr.plot.width=21, repr.plot.height=14)

gams_list %>% 
map(. %>% DimPlot(., reduction = "umap", group.by = 'anno_300cells', pt.size = 1, dims = c(2,3))) %>% 
plot_grid(plotlist = ., labels = names(gams_list), ncol = 2) 
# save_plot("../plots/all_msc_rnagn_count.png", plot = ggplot2::last_plot(), base_height = 6.71, base_asp = 1.618)

# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
# ggplot2::ggsave(filename = "feature.png")
```

```{r}
# options(repr.plot.width=21, repr.plot.height=10)
# DoHeatmap(gams, features = top10$gene_id) + NoLegend()
# # options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID) %>% str_replace(., '_', '-') %>% length
```

```{r}
# gams@assays$RNA@scale.data[stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID) %>% str_replace(., '_', '-'),]
```

Display Rbrewer coloours

```{r}
# brewer.pal.info
```

```{r}
# gams_annot <- gams@meta.data %>% dplyr::select(anno_300cells)
```

```{r}
sorted_gam_scale_data_ffirst_bcodes_list <- gams_list %>% 
map(. %>%
    .@meta.data  %>% 
    dplyr::select(anno_300cells, rank2) %>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(anno_300cells, ~factor(.,levels = c("Gametocytes (comitted)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)"))),
           phmap_colors = case_when(
            anno_300cells == 'Gametocytes (comitted)' ~ brewer.pal(3,"Greys")[1],
            anno_300cells == 'Gametocytes (developing)'~ brewer.pal(3,"Greys")[2],
            anno_300cells == 'Gametocytes (branching)'~ brewer.pal(3,"Greys")[3],
            anno_300cells == 'Gametocytes (early female)'~ brewer.pal(3,"Oranges")[2:3][1],
            anno_300cells == 'Gametocytes (late female)'~ brewer.pal(3,"Oranges")[2:3][2],
            anno_300cells == 'Gametocytes (early male)'~ brewer.pal(3,"Greens")[2:3][1],
            anno_300cells == 'Gametocytes (late male)'~ brewer.pal(3,"Greens")[2:3][2] ),
           across(phmap_colors, ~factor(.,levels = c(brewer.pal(3,"Greys")[1], brewer.pal(3,"Greys")[2], brewer.pal(3,"Greys")[3], brewer.pal(3,"Oranges")[2:3][1], brewer.pal(3,"Oranges")[2:3][2], brewer.pal(3,"Greens")[2:3][1], brewer.pal(3,"Greens")[2:3][2]))),
          ) %>%
    arrange(anno_300cells,rank2)
    )
```

```{r}
# sorted_gam_scale_data_ffirst_bcodes_list[[1]]

```

```{r}
# gams_list
```

```{r}
sorted_gam_scale_data_ffirst_list<- list(gams_list, sorted_gam_scale_data_ffirst_bcodes_list) %>%
pmap(
    ~ ..1@assays$RNA@scale.data %>%
    as.data.frame() %>%
    rownames_to_column('barcode') %>%
    mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
    dplyr::select(barcode, rownames(..2)) %>%
    column_to_rownames('barcode') %>%
    as.matrix()
)
```

```{r}
dr_sex_col <- brewer.pal(3,"Set1")
names(dr_sex_col) <- c('Male', 'Female', 'Female,Male')
```

pheatmap of drivers in gam branch per dataset

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)

list(sorted_gam_scale_data_ffirst_list, sorted_gam_scale_data_ffirst_bcodes_list) %>%
pmap(
    ~ ..1[stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID),] %>%
    pheatmap(
        .,
        cluster_cols = F,
        show_rownames = F,
        # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
        annotation_row = (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% dplyr::select(CDgam_sex, `Gene ID`) %>% column_to_rownames('Gene ID'))%>% rename('Sex correlation' = CDgam_sex),
        annotation_col = ..2 %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        annotation_colors = list(Gametocytes = ..2 %>% dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%deframe, `Sex correlation` = dr_sex_col),
        fontsize = 20
    )
)
    

# options(repr.plot.width=7, repr.plot.height=7)
```

pheatmap of drivers in Committed, developing clusters per dataset

```{r}
options(repr.plot.width=12, repr.plot.height=7)

list(sorted_gam_scale_data_ffirst_list, sorted_gam_scale_data_ffirst_bcodes_list) %>%
pmap(
    ~ ..1[stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% pull(ID),] %>%
    pheatmap(
        .,
        cluster_cols = F,
        show_rownames = F,
        # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
        annotation_row = stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% column_to_rownames('Gene ID') %>% dplyr::select(Bgam_sex) %>% rename('Sex correlation' = Bgam_sex),
        annotation_col = ..2 %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        annotation_colors = list(Gametocytes = ..2 %>% dplyr::count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%deframe, `Sex correlation` = dr_sex_col),
        fontsize = 20
    )
)
    

# options(repr.plot.width=7, repr.plot.height=7)
```



```{r}
# vkg_d10_d10b <- read_csv('data/processed/RoadtoUMAP/var_kine_bf2d10_d10b.csv')

# vkgd10b <- read_csv('data/processed/RoadtoUMAP/var_kine_bf2day10b.csv')

# vkgd10 <- read_csv('data/processed/RoadtoUMAP/var_kine_bf2day10.csv')

# vkgd5 <- read_csv('data/processed/RoadtoUMAP/var_kine_bf2day5.csv')
```

```{r}
library(VennDiagram)
```

```{r}
# vkg_all<-list.files("data/processed/RoadtoUMAP/", pattern = "var_kine_bf2", full.names = T, recursive = T, include.dirs = T)%>% 
#     set_names(nm = (str_remove_all(.,'data/processed/RoadtoUMAP//var_kine_bf2|.csv')  %>% tools::file_path_sans_ext()))%>% 
#     map_df(read_csv, col_types = 'ccdcc', .id = "day")%>% 
# filter(fit_pval_kinetics<0.0001)%>%
# pivot_wider(names_from = day, values_from = fit_pval_kinetics) 
```

```{r}
# Helper function to display Venn diagram
# display_venn <- function(x, ...){
#   library(VennDiagram)
#   grid.newpage()
#   venn_object <- venn.diagram(x, filename = NULL, ...)
#   grid.draw(venn_object)
# }
```

```{r}
# library(VennDiagram) 
# venn.diagram(list(B = 1:1800, A = 1571:2020), fill = c("lightblue", "green"), 
#              alpha = c(0.5, 0.5), lwd =0, "plots/venn_diagram.tiff")
```

```{r}

# ef_scv_stochastic %>%
# map(. %>% .$var %>% 
#     pull(`Gene_ID`)) %>%
# display_venn(., fill = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"), 
#              alpha = c(0.5, 0.5, 0.5, 0.5), lwd =2,  sub.cex = 10, main.cex = 10, cex = 2,cat.cex = 2,
#         cat.fontface = "bold")
```

```{r}
##All DAYS

# options(repr.plot.width=7, repr.plot.height=7)
ef_scv_stochastic %>%
map(. %>% .$var %>% 
    pull(`Gene_ID`)) %>%
ggvenn(.,
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, 
       set_name_size = 9, 
       text_size = 11,
      show_percentage = F)+
scale_x_continuous(expand = expansion(mult = .14))+
scale_y_continuous(expand = expansion(mult = .065))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
# ef_scv_stochastic %>%
# map(. %>% .$var %>% 
#     pull(`Gene_ID`)) %>%
# ggvenn(., fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
#        stroke_size = 0.5, set_name_size = 4)
```

```{r}
QCd_gene_4vel_unique <-ef_scv_stochastic[2:5] %>%
map(. %>% .$var %>% 
    pull(`Gene_ID`)) %>%
unlist() %>%
unique() 
```

```{r}
ef_scv_stochastic[2:5] %>%
map(. %>% .$var %>% 
    pull(`Gene_ID`) %>%
length)
```


```{r}
vkg_all<-ef_scv_stochastic %>%
     map(. %>% .$var %>%
             filter(fit_pval_kinetics<0.0001))%>%
             bind_rows(., .id = "day") %>% dplyr::select( Gene,  `Gene ID`, day, fit_pval_kinetics)%>%
     pivot_wider(names_from = day, values_from = fit_pval_kinetics) %>%
  dplyr::rename('Gene ID'=`Gene ID`)
```

```{r}
# vkg_all %>% filter(if_all(c('all_days', 'd10_d10b', 'day10', 'day10b', 'day5'), ~ .<0.00001))
```

```{r}
vkg_all  %>% dplyr::count(is.na(all_days), is.na(d10_d10b), is.na(day10), is.na(day10b), is.na(day5))
```

```{r}
ggvenn(list(d10_d10b = vkg_all %>%  filter(!is.na(d10_d10b)) %>% pull(Gene_ID),
            day10 = vkg_all %>%  filter(!is.na(day10)) %>% pull(Gene_ID),
            day10b = vkg_all %>%  filter(!is.na(day10b)) %>% pull(Gene_ID),
            day5 = vkg_all %>%  filter(!is.na(day5)) %>% pull(Gene_ID)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
ggvenn(list(day10 = vkg_all %>%  filter(!is.na(day10)) %>% pull(Gene_ID),
            day10b = vkg_all %>%  filter(!is.na(day10b)) %>% pull(Gene_ID),
            day5 = vkg_all %>%  filter(!is.na(day5)) %>% pull(Gene_ID)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 9, text_size=7
      )
```

```{r}
# ,
#        fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
#        stroke_size = 0.5, 
#        set_name_size = 9, 
#        text_size = 11,
#       show_percentage = F)+
# scale_x_continuous(expand = expansion(mult = .14))+
# scale_y_continuous(expand = expansion(mult = .065))
# # options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
ggvenn(list(all_days = vkg_all %>%  filter(!is.na(all_days)) %>% pull(Gene_ID),
            day10 = vkg_all %>%  filter(!is.na(day10)) %>% pull(Gene_ID),
            day10b = vkg_all %>%  filter(!is.na(day10b)) %>% pull(Gene_ID),
            day5 = vkg_all %>%  filter(!is.na(day5)) %>% pull(Gene_ID)
            
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, 
       set_name_size = 9, 
       text_size = 11,
      show_percentage = F)+
scale_x_continuous(expand = expansion(mult = .13))+
scale_y_continuous(expand = expansion(mult = .06))
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
# list(       day10 = vkg_all %>%  filter(!is.na(day10)) %>% pull(Gene_ID),
#             day10b = vkg_all %>%  filter(!is.na(day10b)) %>% pull(Gene_ID),
#             day5 = vkg_all %>%  filter(!is.na(day5)) %>% pull(Gene_ID)
#            )%>%
# display_venn(., fill = c("#0073C2FF", "#EFC000FF", "#868686FF"), 
#              alpha = c(0.5, 0.5, 0.5), lwd =2,  sub.cex = 10, main.cex = 10, cex = 2,cat.cex = 2,
#         cat.fontface = "bold")
```

```{r}
# vkg_all%>%
# group_by(`Gene ID`) %>%
#     top_n(n = 1, wt = fit_pval_kinetics)
```

```{r}
Bunnick_RBPs <- readxl::read_excel("data/raw/published/bunnick_etal/13059_2016_1014_MOESM2_ESM.xlsx", sheet = "RBPs")
```

Reading in DOZI and CITH KO differentially expressed genes

```{r}
CITH_KO <- readxl::read_excel("data/raw/published/Mair_et_al_DOZI/ppat.1000767.s018.xls", skip = 12)
```

```{r}
DOZI_CITH_KO <- readxl::read_excel("data/raw/published/Mair_et_al_DOZI/ppat.1000767.s019.xls", skip = 12)
```

```{r}
DOZI_CITH_IP <- readxl::read_excel("data/raw/published/Mair_et_al_DOZI/ppat.1000767.s017.xls", skip = 13)
```

Preprocessing the old Pf IDs before merging to DOZI and CITH tables with Pf orthologs

```{r}
pf_old_ids_pp <- Pf3D7_genes_plasmoDB %>% 
dplyr::select(gene_id, `Gene Name or Symbol`, `Previous ID(s)`) %>% 
separate(`Previous ID(s)`, c('tbd', 'mal_id', 'pf_id','xtr', 'xtr2', 'xtr3','xtr4','xtr5','xtr6'), ';|:') %>%
mutate(across(c('tbd', 'mal_id', 'pf_id','xtr', 'xtr2', 'xtr3','xtr4','xtr5','xtr6'), ~trimws(.))) %>% 
pivot_longer(cols = c('mal_id','pf_id','xtr','xtr2','xtr3','xtr4','xtr5','xtr6'), names_to = 'id_category', values_drop_na=T, values_to = 'all_old_ids')
```

DOZI and CITH

```{r}
DOZI_CITH_KO_new_PfIDs <- DOZI_CITH_KO %>% 
filter(if_any(everything(), ~ !is.na(.))) %>% 
left_join(., pf_old_ids_pp %>% dplyr::select(gene_id, `Gene Name or Symbol`, all_old_ids), by = c('Pf orthologs'='all_old_ids'))  
```

```{r}
# DOZI_CITH_KO_new_PfIDs %>% filter(`WT vs DOZI-KO` > 0 & `WT vs CITH-KO` > 0) %>% pull(gene_id) %>% str_replace('-', '_')
```

```{r}

# ef_scv_stochastic %>%
# map(. %>% .$var %>% 
#     pull(`Gene ID`)) %>%
# display_venn(., fill = c("#0073C2FF", "#EFC000FF", "#868686FF"), 
#              alpha = c(0.5, 0.5, 0.5), lwd =2,  sub.cex = 10, main.cex = 10, cex = 2,cat.cex = 2,
#         cat.fontface = "bold")
```

```{r}
# list('DOZI & CITH KO DR' = DOZI_CITH_KO_new_PfIDs %>% filter(`WT vs DOZI-KO` > 0 & `WT vs CITH-KO` > 0) %>% pull(gene_id) %>% str_replace('-', '_'),
#             female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB)  & CDgam_sex == 'Female') %>% pull(ID) ,
#             female_upregulated = fVSm %>% filter(avg_log2FC>0) %>% rownames() %>% str_replace(., '-', '_'),
#             'Genes with variable kinetics' = vkg_all %>% pull(ID)
#            )%>%
# display_venn(., fill = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"), 
#              alpha = c(0.5, 0.5, 0.5,1), lwd =2,  cex = 2,cat.cex = 2,cat.fontface = "bold")
```

```{r}
ggvenn(list('DOZI & CITH KO DR' = DOZI_CITH_KO_new_PfIDs %>% filter(`WT vs DOZI-KO` > 0 & `WT vs CITH-KO` > 0) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB)  & CDgam_sex == 'Female') %>% pull(ID) ,
            female_upregulated = fVSm %>% filter(avg_log2FC>0) %>% rownames() %>% str_replace(., '-', '_'),
            'Genes with variable kinetics' = vkg_all %>% pull(ID)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.7, set_name_size = 5, text_size = 5
      )
```

```{r}
vkg_all %>% filter(ID %in% (DOZI_CITH_KO_new_PfIDs %>% filter(`WT vs DOZI-KO` > 0 & `WT vs CITH-KO` > 0) %>% pull(gene_id) %>% str_replace('-', '_')))
```

DOZI only

```{r}
CITH_KO_new_PfIDs <- CITH_KO %>% 
filter(if_any(everything(), ~ !is.na(.))) %>% 
left_join(., pf_old_ids_pp %>% dplyr::select(gene_id, `Gene Name or Symbol`, all_old_ids), by = c('Pf ortholog'='all_old_ids')) 
```

```{r}
vkg_all %>% filter(`Gene_ID` %in% (CITH_KO_new_PfIDs %>% filter(`WT vs CITH` > 0 ) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID)
```

```{r}
ggvenn(list('CITH KO DR' = CITH_KO_new_PfIDs %>% filter(`WT vs CITH` > 0) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),
            female_upregulated = fVSm %>% filter(avg_log2FC>0) %>% rownames() %>% str_replace(., '-', '_'),
            'Genes with variable kinetics' = vkg_all %>% pull(`Gene_ID`)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
 
ggvenn(list('CITH KO DR' = CITH_KO_new_PfIDs %>% filter(`WT vs CITH` > 0) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),
            female_dr_gambranch = stocha_pk_gns %>% filter(!is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),
            'Genes with variable kinetics' = vkg_all %>% pull(`Gene_ID`)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
ggvenn(list('CITH KO DR' = CITH_KO_new_PfIDs %>% filter(`WT vs CITH` > 0) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_gambranch = stocha_pk_gns %>% filter(!is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),
            female_upregulated = fVSm %>% filter(avg_log2FC>0) %>% rownames() %>% str_replace(., '-', '_'),
            'Genes with variable kinetics' = vkg_all %>% pull(`Gene_ID`)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
# DOZI_CITH_IP_new_PfIDs <- DOZI_CITH_IP %>% 
# filter(if_any(everything(), ~ !is.na(.))) %>% 
# left_join(., pf_old_ids_pp %>% dplyr::select(gene_id, `Gene Name or Symbol`, all_old_ids), by = c('Pf orthologs'='all_old_ids')) 
```

```{r}
DOZI_CITH_IP_new_PfIDs <- DOZI_CITH_IP %>% 
filter(if_any(everything(), ~ !is.na(.)))%>% 
left_join(., pf_old_ids_pp %>% dplyr::select(gene_id, `Gene Name or Symbol`, all_old_ids), by = c('protein name...16'='all_old_ids')) 
```

```{r}
DOZI_CITH_IP_new_PfIDs %>% filter(`av ratio` > 3)
```

```{r}
ggvenn(list('CITH KO IP' = DOZI_CITH_IP_new_PfIDs %>% filter(`av ratio` > 3) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),
            female_dr_gambranch = stocha_pk_gns %>% filter(!is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),
            'Genes with variable kinetics' = vkg_all %>% pull(`Gene_ID`)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
ggvenn(list('CITH KO IP' = DOZI_CITH_IP_new_PfIDs %>% filter(`av ratio` > 3) %>% pull(gene_id) %>% str_replace('-', '_'),
            female_dr_CDB = stocha_pk_gns %>% filter(!is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),
            female_dr_gambranch = stocha_pk_gns %>% filter(!is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),
            'Genes with variable kinetics' = vkg_all %>% pull(`Gene_ID`)
           ),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
       stroke_size = 0.5, set_name_size = 4
      )
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.)))
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% filter(`av ratio` > 3) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% filter(`ratio...5` > 3) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% filter(`ratio...8` > 3) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% filter(`ratio...11` > 3) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
vkg_all %>% filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% filter(`ratio...11` > 3) %>% pull(gene_id) %>% str_replace('-', '_')))
```

```{r}
# vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% 
# filter(`Gene_ID` %in% (DOZI_CITH_IP_new_PfIDs %>% pull(gene_id) %>% str_replace('-', '_')))%>% 
# mutate(across(`...1`, factor)) %>% 
# group_by(`...1`) 
```

```{r}
# vkg_all %>% 
# distinct(`Gene ID`, .keep_all = T) %>%
# left_join(., Bunnick_RBPs, by = c('Gene ID' = 'GeneID')) 
```

```{r}
# vkg_all %>% 
# left_join(., Bunnick_RBPs, by = c('Gene ID' = 'GeneID')) %>% 
# dplyr::select(fit_diff_kinetics, Classification, Function, Evidence,fit_pval_kinetics) %>% 
# mutate(variable_kinetics_sex = case_when(str_detect(fit_diff_kinetics, ' male') ~ 'Male',
#                                         str_detect(fit_diff_kinetics, ' female') ~ 'Female',
#                                         str_detect(fit_diff_kinetics, 'comitted|branching|developing') ~ 'CBD')) %>%
# count(variable_kinetics_sex, Evidence)
```

```{r}
vkg_all %>% head
```

```{r}
vkg_proc <-ef_scv_stochastic %>%
    map(. %>% .$var %>%
            filter(fit_pval_kinetics<0.0001))%>%
    bind_rows(., .id = "day") %>% dplyr::select( Gene,  `Gene ID`, day, fit_pval_kinetics, fit_diff_kinetics)%>% 
    distinct(`Gene ID`, .keep_all = T) %>%
    left_join(., Bunnick_RBPs, by = c('Gene ID' = 'GeneID')) %>% 
    dplyr::select(Gene_ID, fit_diff_kinetics, Classification, Function, Evidence) %>% 
mutate(variable_kinetics_sex = case_when(str_detect(fit_diff_kinetics, ' male') ~ 'Male',
                                        str_detect(fit_diff_kinetics, ' female') ~ 'Female',
                                        str_detect(fit_diff_kinetics, 'comitted|branching|developing') ~ 'CBD')) 
```

```{r}
# vkg_proc <- vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.)))  %>% 
# distinct(`Gene ID`, .keep_all = T) %>%
# left_join(., Bunnick_RBPs, by = c('Gene ID' = 'GeneID')) %>% 
# dplyr::select(Gene_ID, fit_diff_kinetics, Classification, Function, Evidence) %>% 
# mutate(variable_kinetics_sex = case_when(str_detect(fit_diff_kinetics, ' male') ~ 'Male',
#                                         str_detect(fit_diff_kinetics, ' female') ~ 'Female',
#                                         str_detect(fit_diff_kinetics, 'comitted|branching|developing') ~ 'CBD')) 
```

```{r}
vkg_proc
```

```{r}
vkg_proc %>% filter(variable_kinetics_sex == 'Female') 
```

```{r}
# vkg_proc_high_conf <- vkg_proc %>% filter(fit_pval_kinetics < 0.000005)
```

```{r}
# vkg_proc_high_conf%>%
# count(variable_kinetics_sex, Evidence)
```

```{r}
# options(repr.plot.width=15, repr.plot.height=16)
# sorted_gam_scale_data[vkg_proc_high_conf %>% pull(ID),] %>%
# pheatmap(., 
#          cluster_cols = F, 
#          # show_rownames = F,
#          labels_row = PfPDB_genes%>% filter(gene_id %in% (vkg_proc_high_conf %>% pull(ID)))  %>% pull(Gene),
#         show_colnames = F,
#          annotation_row = vkg_proc_high_conf %>% as.data.frame() %>% column_to_rownames('Gene ID') %>% dplyr::select(variable_kinetics_sex),
#         annotation_col = gams@meta.data %>% dplyr::select(anno_300cells))
# # options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
org.Pf.plasmo.db
```

```{r}
geneList = keys(org.Pf.plasmo.db)
```

```{r}
geneList %>% length
```

```{r}
##with list of genes qcd for rna velocity
```

```{r}
geneList[geneList %in% QCd_gene_4vel_unique] %>% length
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
# qcd_vel_genes <- geneList()
```

```{r}
dr_cdb_enr <- enrichGO(stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP")
```

```{r}
##plotting
plot <- dotplot(dr_cdb_enr,font.size=8,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_CDB) & CDgam_sex == 'Male') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP")
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_CDB) & CDgam_sex == 'Male') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_CDB) & CDgam_sex == 'Male') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") 
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(female_drivers_in_CDB) & CDgam_sex == 'Female') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_gambranch) & Bgam_sex == 'Male') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_gambranch) & Bgam_sex == 'Male') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(male_drivers_in_gambranch) & Bgam_sex == 'Male') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
goplot(.)
```

```{r}
# dotplot(kk2, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(stocha_pk_gns %>% filter( !is.na(female_drivers_in_gambranch) & Bgam_sex == 'Female') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(Bgam_annot %>%
         filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
         filter(SexDE == 'Male') %>%
         rownames(),
         universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', 
         keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(Bgam_annot %>%
         filter(SexDE != 'Not DE' & `Sex correlation` == SexDE) %>%
         filter(SexDE == 'Male') %>%
         rownames(),
         universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', 
         keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
library(DOSE)
data(geneList)
de <- names(geneList)[abs(geneList) > 2]
```

```{r}
# geneList
```

```{r}
# vkg_proc_high_conf %>% filter(variable_kinetics_sex == 'Female') %>% pull(ID)
```

```{r}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.)))
```

```{r, results="hide"}
vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) 
```

```{r}
enrichGO(vkg_all %>% filter(!if_all(c('day10','day10b','day5'), ~is.na(.))) %>% filter(variable_kinetics_sex == 'Female') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP")
```

```{r}
enrichGO(vkg_proc %>% filter(variable_kinetics_sex == 'Female') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP")
```

```{r, fig.width= 6, fig.height=6}
options(repr.plot.width=6, repr.plot.height=6)
enrichGO(vkg_proc %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + 
theme(legend.key.height = unit(2,'mm'),
      axis.text=element_text(size=40), 
              axis.title=element_text(size=20,face="bold"), 
              legend.text = element_text(size=10), 
              legend.title = element_text(size=15,face="bold"),
                strip.text = element_text(size = 20),
             title = element_text(size = 15))+
          rotate_x_text()

# options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
enrichGO(vkg_proc %>% filter(variable_kinetics_sex == 'CBD') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + 
theme(legend.key.height = unit(2,'mm')) +
        theme(axis.text=element_text(size=40), 
              axis.title=element_text(size=20,face="bold"), 
              legend.text = element_text(size=10), 
              legend.title = element_text(size=15,face="bold"),
                strip.text = element_text(size = 20),
             title = element_text(size = 15)) +
          rotate_x_text()
```

```{r}
enrichGO(vkg_proc_high_conf %>% filter(variable_kinetics_sex == 'CBD') %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
# enrichGO(vkg_proc_high_conf  %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
# dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(vkg_proc %>% filter(variable_kinetics_sex == 'Female') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
# enrichGO(vkg_proc_high_conf %>% filter(variable_kinetics_sex == 'CBD') %>% pull(ID),universe=geneList[geneList %in% QCd_gene_4vel_unique], OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
# dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
enrichGO(vkg_proc %>% pull(ID),universe=geneList, OrgDb='org.Pf.plasmo.db', keyType ="SYMBOL",ont="BP") %>%
dotplot(.,font.size=14,showCategory=4) + theme(legend.key.height = unit(2,'mm'))
```

```{r}
#get velocity for velocity genes (used to calculate velocities (Genes for which the velocity values in all cells are not equal to zero may be an imperfect  a subset of this))
bf2_vels_4velogns <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
pmap(~ ..1$layers[['velocity']] %>%
    `colnames<-` (..2) %>%
    `rownames<-` (..3) %>%
     as.data.frame() %>%
    dplyr::select((..1$var %>% 
           filter(velocity_genes == TRUE) %>% 
           pull(Gene_ID)))
    )
```

```{r}
# #get velocity for velocity genes (used to calculate velocities (Genes for which the velocity values in all cells are not equal to zero may be an imperfect  a subset of this))
# bf2_u_4velogns <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
# pmap(~ ..1$layers[['unspliced']] %>%
#     `colnames<-` (..2) %>%
#     `rownames<-` (..3) %>%
#      as.data.frame() %>%
#     dplyr::select((..1$var %>% 
#            filter(velocity_genes == TRUE) %>% 
#            pull(Gene_ID)))
#     )
```

```{r}
D5_bf2_genes %>% length()
```

```{r}
#get spliced for all QCd genes and scale
bf2_s_all_scaled <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
pmap(~ ..1$layers[['spliced']] %>%
    `colnames<-` (..2) %>%
    `rownames<-` (..3) %>%
     as.matrix() %>% 
     t()%>% 
     FastRowScale()
    )
```

```{r}
#get unspliced for all QCd genes and scale
bf2_u_all_scaled <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
pmap(~ ..1$layers[['unspliced']] %>%
    `colnames<-` (..2) %>%
    `rownames<-` (..3) %>%
     as.matrix() %>% 
     t()%>% 
     FastRowScale()
    )
```

```{r}
# D5_bf2_u_velg_scaled <- D5_bf2_u %>% as.data.frame() %>% dplyr::select(all_of(D5_bf2_vel.genes)) %>% as.matrix() %>% t() %>% FastRowScale()
```

```{r}
#  ef_qcd_bf2_scv_D5$obs %>% 
# dplyr::select(anno_300cells, rank2)
```

```{r}
gam_factors = c("Gametocytes (comitted)", "Gametocytes (developing)", "Gametocytes (branching)", "Gametocytes (early female)", "Gametocytes (late female)", "Gametocytes (early male)", "Gametocytes (late male)")
```

```{r}
bf2_sorted_gam_scale_data_ffirst_bcodes <- ef_scv_stochastic %>% 
map(. %>%
    .$obs %>% 
    dplyr::select(anno_300cells, rank2) %>% 
    mutate(across(rank2, ~as.numeric(.)),
           across(anno_300cells, ~factor(.,levels = gam_factors))) %>%
    arrange(anno_300cells,rank2) 
    )
```

```{r}
bf2_sorted_gam_scale_data_ffirst <- list(bf2_u_all_scaled, bf2_sorted_gam_scale_data_ffirst_bcodes) %>% 
pmap(~ ..1 %>%
    as.data.frame() %>%
    rownames_to_column('barcode') %>%
    mutate(across(barcode, ~str_replace(., '-', '_'))) %>%
    dplyr::select(barcode, ..2 %>% rownames()) %>%
    column_to_rownames('barcode') %>%
    as.matrix()
   )
```

```{r}
# sorted_gam_scale_data_ffirst_bcodes %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes)
```

```{r}
# options(repr.plot.width=12, repr.plot.height=7)

# list(sorted_gam_scale_data_ffirst_list, sorted_gam_scale_data_ffirst_bcodes_list) %>%
# pmap(
#     ~ ..1[stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% pull(ID),] %>%
#     pheatmap(
#         .,
#         cluster_cols = F,
#         show_rownames = F,
#         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
#         show_colnames = F,
#         annotation_row = stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% column_to_rownames('Gene ID') %>% dplyr::select(Bgam_sex) %>% rename('Sex correlation' = Bgam_sex),
#         annotation_col = ..2 %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
#         annotation_colors = list(Gametocytes = ..2 %>% count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%deframe, `Sex correlation` = dr_sex_col),
#         fontsize = 20
#     )
# )
    

# # options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
ef_scv_stochastic[[1]]$obs %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes)
```

```{r}
# options(repr.plot.width=12, repr.plot.height=7)

# list(bf2_sorted_gam_scale_data_ffirst, ef_scv_stochastic) %>%
# pmap(
#     ~ ..1 %>%
#     pheatmap(., 
#          cluster_cols = F, 
#          show_rownames = F,
#          # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
#         show_colnames = F,
#          # annotation_row = (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% dplyr::select(CDgam_sex, `Gene ID`) %>% column_to_rownames('Gene ID'))%>% rename('Sex correlation' = CDgam_sex),
#         annotation_col = ..2$obs %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
#         fontsize = 20)
#     )
# # options(repr.plot.width=7, repr.plot.height=7)
```

```{r}
bf2_genes_with_switch_tps <- ef_scv_stochastic %>% 
map(. %>%
    .$var %>% 
    dplyr::select(Gene_ID, 'fit_t_',fit_likelihood) %>% 
    filter(fit_t_ != 'NaN') %>% 
    arrange(fit_t_) %>% 
    rownames_to_column('bcode') %>%
    mutate(across(`Gene ID`, ~case_when(. == 'selfdefinedNA' ~ bcode,
                                    TRUE ~ as.character(.))))
    )
```

```{r}
bf2_us_scaled <- list(ef_scv_stochastic,bf2_genes,bf2_cells)  %>% 
pmap(~ ..1$layers[['unspliced']] %>%
    `colnames<-` (..2) %>%
    `rownames<-` (..3) %>%
    as.data.frame() %>% 
    rownames_to_column('bcode') %>% 
    mutate(across(bcode,~paste(.,'U', sep = '_'))) %>%
    rbind(., ..1$layers[['spliced']] %>%
          `colnames<-` (..2) %>%
          `rownames<-` (..3) %>% 
          as.data.frame() %>% 
          rownames_to_column('bcode') %>% 
          mutate(across(bcode,~paste(.,'S', sep = '_')))) %>%
    column_to_rownames('bcode') %>% 
    as.matrix() %>% 
    t() %>%
    FastRowScale()
     )
```

```{r}


# D5_bf2_us_scaled <- D5_bf2_u %>% 
# as.data.frame() %>% 
# rownames_to_column('bcode') %>% 
# mutate(across(bcode,~paste(.,'U', sep = '_'))) %>%
# rbind(.,D5_bf2_s %>% 
#       as.data.frame() %>% 
#       rownames_to_column('bcode') %>% 
#       mutate(across(bcode,~paste(.,'S', sep = '_')))) %>%
# column_to_rownames('bcode') %>% 
# as.matrix() %>% 
# t() %>%
# FastRowScale()
```

```{r}
# D5_bf2_us_scaled
```

```{r}
bf2_sorted_gam_scale_data_ffirst_bcodes_us <- list(bf2_sorted_gam_scale_data_ffirst_bcodes, bf2_sorted_gam_scale_data_ffirst_bcodes ) %>% 
pmap(~ rbind(..1 %>%
             rownames_to_column('bcode') %>% 
             mutate(across(anno_300cells,~paste(.,'U')),
                    across(bcode,~paste(.,'U', sep = '_'))),
             ..2 %>%
             rownames_to_column('bcode') %>% 
             mutate(across(anno_300cells,~paste(.,'S')),
                    across(bcode,~paste(.,'S', sep = '_')))
            ) %>%
     column_to_rownames('bcode') %>%
     dplyr::select(anno_300cells, rank2) %>% 
     mutate(across(rank2, ~as.numeric(.)),
            across(anno_300cells, ~factor(.,levels = c(paste(rep(gam_factors, each=2), c("U", "S")))))) %>%
     rename('Gametocytes' = anno_300cells)
    )
```

```{r}
bf2_sorted_gam_scale_data_ffirst_bcodes_us[[1]]
```

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)

list(bf2_us_scaled, bf2_genes_with_switch_tps, bf2_sorted_gam_scale_data_ffirst_bcodes_us) %>%
pmap(
    ~ ..1[ ..2 %>% filter(fit_likelihood > 0.3) %>% pull(ID), ..3 %>% rownames()] %>%
    pheatmap(., 
         cluster_cols = F, 
         cluster_rows = F,
         show_rownames = F,
         # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
         # annotation_row = (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% dplyr::select(CDgam_sex, `Gene ID`) %>% column_to_rownames('Gene ID'))%>% rename('Sex correlation' = CDgam_sex),
        annotation_col = ..3 %>%dplyr::select(Gametocytes) ,
        fontsize = 20)
    )
# options(repr.plot.width=7, repr.plot.height=7)
```

```{r, fig.width= 12, fig.height=7}
options(repr.plot.width=12, repr.plot.height=7)

list(sorted_gam_scale_data_ffirst_list, bf2_genes_with_switch_tps, sorted_gam_scale_data_ffirst_bcodes_list) %>%
pmap(
    ~ ..1[ ..2 %>% filter(fit_likelihood > 0.3) %>% pull(ID),] %>%
    pheatmap(
        ., 
        cluster_cols = F,
        show_rownames = F,
        cluster_rows = F,
        # labels_row = PfPDB_genes%>% filter(gene_id %in% (stocha_pk_gns %>% filter(!is.na(Bgam_sex_stocha_VKCK))  %>% pull(ID)))  %>% pull(Gene),
        show_colnames = F,
        # annotation_row = stocha_pk_gns %>% filter(if_any(matches('dir.*ale_in_gambranch'), ~.== 'correlated') )  %>% column_to_rownames('Gene ID') %>% dplyr::select(Bgam_sex) %>% rename('Sex correlation' = Bgam_sex),
        annotation_col = ..3 %>% rename('Gametocytes' = anno_300cells) %>%dplyr::select(Gametocytes),
        annotation_colors = list(Gametocytes = ..3 %>% count(anno_300cells, phmap_colors) %>% mutate_if(is.factor, as.character) %>%deframe, `Sex correlation` = dr_sex_col),
        fontsize = 20
    )
    )
# options(repr.plot.width=7, repr.plot.height=7)
        
```

```{r}
D5_bf2_u_all_scaled %>% dim()
```

```{r}
colnames(D5_bf2_curr) <- D5_bf2_genes
rownames(D5_bf2_curr) <- D5_bf2_cells
D5_bf2_curr <- as.matrix(curr)[,D5_bf2_vel.genes]
D5_bf2_curr <- t(D5_bf2_curr)
```

```{r}
#cell type assignments
D5_bf2_clusters <- ef_qcd_bf2_scv_D5$obs$Stage
names(D5_bf2_clusters) <- D5_bf2_cells
```

```{r}
#colors
D5_bf2_col = rev(plasma(length(levels(D5_bf2_clusters))))
D5_bf2_cell.cols = D5_bf2_col[D5_bf2_clusters]
names(D5_bf2_cell.cols) = names(D5_bf2_clusters)
```

```{r}
D5_bf2_col
```

```{r}
#get velocity
D5_bf2_vel <- ef_qcd_bf2_scv_D5$layers[['velocity']]
colnames(D5_bf2_vel) <- D5_bf2_genes
rownames(D5_bf2_vel) <- D5_bf2_cells
```

```{r}
D5_bf2_vel.genes <- D5_bf2_genes[colSums(D5_bf2_vel, na.rm=T)>0]
D5_bf2_vel <- D5_bf2_vel[,D5_bf2_vel.genes]
```

```{r}
#get current
D5_bf2_curr <- ef_qcd_bf2_scv_D5$X
```

```{r}
D5_bf2_vel.genes
```

```{r}
colnames(D5_bf2_curr) <- D5_bf2_genes
rownames(D5_bf2_curr) <- D5_bf2_cells
D5_bf2_curr <- as.matrix(curr)[,D5_bf2_vel.genes]
D5_bf2_curr <- t(D5_bf2_curr)
```

```{r}
#compute projected
D5_bf2_proj <- D5_bf2_curr + t(vel)
D5_bf2_D5_bf2_proj[D5_bf2_proj<0] <- 0
```

```{r}
ef_qcd_bf2_scv_D5$obsm['X_umap']
```

```{r}
scv$pl$velocity_embedding_stream(ef_qcd_bf2_scv_D5, basis='umap', save = 'day5_umap_vviz.png')
```

```{r}
plotEmbedding(ef_qcd_bf2_scv_D5$obsm['X_umap'],
              main = 'UMAP', xlab = "X", ylab = "Y")
```

```{r}
Convert("data/processed/PfDB52e/day5/scv_bf2_b4CR.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r}
a.plot <- adata

#remove unconnected cells from anndata object
connected.cells <-  rownames(emb.veloviz)
a.plot.vv <- a.plot[a.plot$obs_names$isin(connected.cells)]
a.plot.vv$obsm$update("X_veloviz1" = emb.veloviz)
```

```{r}
v3_days = c('asexuals', 'day3', 'day5', 'day10', 'day10b', 'd10_d10b')
```

```{r, results="hide"}
for (v3_day in v3_days){
    Convert(paste0("data/processed/PfDB52e/", v3_day, "/scv_bf1_b4CR.h5ad"), dest = "h5seurat", overwrite = TRUE)
    
}
    
```

```{r, results="hide"}
day5<- LoadH5Seurat('data/processed/PfDB52e/day5/scv_bf1_b4CR.h5seurat', misc = F)
```

```{r}
adata <- day5
```

```{r}
use_condaenv("../envs/scv_jupcon_py3.8/", required = TRUE)
scv <- import("scvelo")
```

```{r}
plt <- import("matplotlib.pyplot") 
```

```{r}
scv$logging$print_version()
```

```{r}
adata <- scv$read('data/processed/PfDB52e/day5/scv_bf1_b4CR.h5ad')
```

```{r}
# jc_dyn_driver_gns %>% 
# rename("rank" = `...1`) %>% 
# pivot_longer(cols = -c("rank"), names_to = "Driver_Stage_JC", values_to = "Gene ID") %>%
# unite("gn_stg_id", Driver_Stage_JC:`Gene ID`, remove = F) %>% 
# left_join(., painter_new_transcripts_0_48, by = "Gene ID") %>%
# group_by(Driver_Stage_JC) %>%
# slice_min(order_by = rank, n = 20) %>%
# pivot_longer(cols = -c('Gene ID', 'Annotation', 'Peak Time (hpi)', 'p-value', 'rank', 'gn_stg_id', 'Driver_Stage_JC'), names_to = "Hours post-invasion") %>% 
# mutate(across(`Hours post-invasion`, ~as.numeric(str_extract(., "[[:digit:]]+")))) %>%
# ggplot(., aes(x = `Hours post-invasion`, color = `Gene ID`, group = `Gene ID`, y = log2(value))) + 
# geom_point() + 
# geom_line() + 
# theme(legend.position = "none") +
# facet_wrap(~ Driver_Stage_JC)

# ggsave("../plots/jcrv_stocha_dr_gns_in_paintr_hpi.png")
```

```{r}
as_velo <- read_csv("data/processed/PfDB52e/asexuals/velos.csv") %>% dplyr::select(-c(1))
```

```{r}
as_cells <- read_lines("data/processed/PfDB52e/asexuals/cells.txt")
```

```{r}
rownames(as_velo) <- as_cells
```

```{r}
names(colSums(as_velo, na.rm=T)>0) %>% length()
```

```{r}
table(colSums(as_velo, na.rm=T)>0 )
```

```{r}
as_velo <- as_velo[,colnames(as_velo)[colSums(as_velo, na.rm=T)>0]]
```

```{r}
as_currX <- read_csv("data/processed/PfDB52e/asexuals/currentX.csv") %>% dplyr::select(-c(1))
```

```{r}
rownames(as_currX) <- as_cells
```

```{r}
as_currX <- as.matrix(as_currX)[,colnames(as_velo)[colSums(as_velo, na.rm=T)>0]]
```

```{r}
as_currX <- t(as_currX)
```

```{r}
as_proj <- as_currX + t(as_velo)
```

```{r}
as_proj[as_proj<0] <- 0
```

```{r, results="hide"}
as_proj
```

```{r}
veloviz = buildVeloviz(
  curr = as_currX, proj = as_proj,
  normalize.depth = TRUE,
  use.ods.genes = TRUE,
  alpha = 0.05,
  pca = TRUE,
  nPCs = 20,
  center = TRUE,
  scale = TRUE,
  k = 20,
  similarity.threshold = 0.2,
  distance.weight = 1,
  distance.threshold = 1,
  weighted = TRUE,
  # seed = 1,
  verbose = FALSE
)
```

```{r}
emb.veloviz <- veloviz$fdg_coords
```

```{r}
as_stages <- read_csv("data/processed/PfDB52e/asexuals/stages.csv")
```

```{r}
as_stages <- as_stages %>% pull('Stage', name = '...1') %>% factor(.)
```

```{r}
levels(as_stages)
```

```{r}
col = rev(plasma(length(levels(as_stages))))
```

```{r}
cell.cols = col[as_stages]
```

```{r}
cell.cols
```

```{r}
names(cell.cols) = names(as_stages)
```

```{r}
plotEmbedding(emb.veloviz, colors = cell.cols[rownames(emb.veloviz)], main='VeloViz with dynamical velocity',
              xlab = "VeloViz X", ylab = "VeloViz Y",
              alpha = 0.8,
              cex.lab = 1.5)
```

```{r}
# Convert("data/processed/PfDB52e/asexuals/scv_b4KE.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r}
# as_scv_b4KE <- LoadH5Seurat("data/processed/PfDB52e/asexuals/scv_b4KE.h5seurat")
```

```{r, results="hide"}
# Convert("data/processed/PfDB52e/asexuals/scv_bf1_b4CR.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r, results="hide"}
# as_ef_qc_py <- LoadH5Seurat("data/processed/PfDB52e/asexuals/scv_bf1_b4CR.h5seurat", misc = F)
```

```{r}
v3_days = c('asexuals', 'day3', 'day5', 'day10', 'day10b', 'd10_d10b')
```

```{r}
# for (v3_day in v3_days){
#     Convert(paste0("data/processed/PfDB52e/", v3_day, "/scv_bf1_b4CR.h5ad"), dest = "h5seurat", overwrite = TRUE)
    
# }
    
```

```{r}
# day5<- LoadH5Seurat('data/processed/PfDB52e/day5/scv_bf1_b4CR.h5seurat', misc = F)
```

```{r}
# day5
```

```{r}
# list.files("data/processed/PfDB52e/", pattern = "scv_bf1_b4CR.h5seurat", full.names = F, recursive = T, include.dirs = T)
```

```{r}
# list.files("data/processed/PfDB52e/", pattern = "scv_bf1_b4CR.h5seurat", full.names = T, recursive = T, include.dirs = T)
```

```{r}
# str_extract('data/processed/PfDB52e//asexuals/scv_bf1_b4CR.h5seurat', 'PfDB52e//*/scv_')
```

```{r}
# str_extract('data/processed/PfDB52e//asexuals/scv_bf1_b4CR.h5seurat',"(?<=PfDB52e////)(.+)(?=\\scv_)")
```

```{r}
# scv_bf1_b4CR_lst <- list.files("data/processed/PfDB52e/", pattern = "scv_bf1_b4CR.h5seurat", full.names = T, recursive = T, include.dirs = T) %>%
#     str_sort(., numeric = T)  %>%
#     map(~LoadH5Seurat(., misc = F)) %>% 
#     set_names()
# #     map2(., v3_days, ~ CreateSeuratObject(.x, project = paste0(.y, ".mrna")))
    
```

```{r}
# scv_bf1_b4CR_lst
```

```{r}
# list.files("data/processed/PfDB52e/", pattern = "scv_bf2_b4CR.h5seurat", full.names = F, recursive = T, include.dirs = T)
```

```{r}
# scv_bf1_b4CR_lst <- list.files("data/processed/PfDB52e/", pattern = "scv_bf1_b4CR.h5seurat", full.names = T, recursive = T, include.dirs = T) %>%
#     str_sort(., numeric = T)  %>%
#     map(~LoadH5Seurat(., misc = F)) %>% 
#     set_names(., )
```

```{r}
# list = []
```

```{r}
# list.files("data/raw/PfDB52e/", pattern = "filtered_feature_bc_matrix$", full.names = T, recursive = T, include.dirs = T) %>% 
#     str_sort(., numeric = T) %>%
```

```{r}

# for (v3_day in v3_days){
#     Convert(paste0("data/processed/PfDB52e/", v3_day, "/scv_bf1_b4CR.h5ad"), dest = "h5seurat", overwrite = TRUE)
    
# }
```

```{r}
# as_ef_qc_py
```

```{r}
# Convert("data/processed/PfDB52e/asexuals/scv_bf2_b4CR.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r}
# as_ef_qc_py <- LoadH5Seurat("data/processed/PfDB52e/asexuals/scv_bf2_b4CR.h5seurat")
```

```{r}
# Convert("data/processed/RoadtoUMAP/ef_qcd_bf2.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r}
# Convert("data/processed/PfDB52e/asexuals/ef_qc.h5ad", dest = "h5seurat", overwrite = TRUE)
```

```{r}
# as_ef_qc_py <- LoadH5Seurat("data/processed/PfDB52e/asexuals/ef_qc.h5seurat")
```

```{r}
use_condaenv("../envs/scv_jupcon_py3.8/", required = TRUE)
scv <- import("scvelo")
```

```{r}
scv$logging$print_version()
```

```{r}
scv$logging$print_version()
```

```{r}
adata <- scv$datasets$pancreas()
```

```{r}
a.plot <- adata

#remove unconnected cells from anndata object
connected.cells <-  rownames(emb.veloviz)
a.plot.vv <- a.plot[a.plot$obs_names$isin(connected.cells)]
a.plot.vv$obsm$update("X_veloviz1" = emb.veloviz)
```

```{r}
v3_days = c('asexuals', 'day3', 'day5', 'day10', 'day10b', 'd10_d10b')
```

```{r, results="hide"}
for (v3_day in v3_days){
    Convert(paste0("data/processed/PfDB52e/", v3_day, "/scv_bf1_b4CR.h5ad"), dest = "h5seurat", overwrite = TRUE)
    
}
    
```

```{r, results="hide"}
day5<- LoadH5Seurat('data/processed/PfDB52e/day5/scv_bf1_b4CR.h5seurat', misc = F)
```

```{r}
adata <- day5
```

```{r}
use_condaenv("../envs/scv_jupcon_py3.8/", required = TRUE)
scv <- import("scvelo")
```

```{r}
plt <- import("matplotlib.pyplot") 
```

```{r}
scv$logging$print_version()
```

```{r}
adata <- scv$read('data/processed/PfDB52e/day5/scv_bf1_b4CR.h5ad')
```

```{r}
#gene and cell names
genes <- adata$var_names$values
cells <- adata$obs_names$values
```

```{r}
#cell type assignments
clusters <- adata$obs$Stage
names(clusters) <- cells
```

```{r}
#colors
col = rev(plasma(length(levels(clusters))))
cell.cols = col[clusters]
names(cell.cols) = names(clusters)
```

```{r}
col
```

```{r}
#get velocity
vel <- adata$layers[['velocity']]
colnames(vel) <- genes
rownames(vel) <- cells
```

```{r}
vel.genes <- genes[colSums(vel, na.rm=T)>0]
vel <- vel[,vel.genes]
```

```{r}
#get current
curr <- adata$X
```

```{r}
vel.genes
```

```{r}
colnames(curr) <- genes
rownames(curr) <- cells
curr <- as.matrix(curr)[,vel.genes]
curr <- t(curr)
```

```{r}
#compute projected
proj <- curr + t(vel)
proj[proj<0] <- 0
```

```{r}
adata
```

```{r, results="hide"}
adata$obsm['X_umap']
```

```{r}
scv$pl$velocity_embedding_stream(adata, basis='umap', save = 'day5_umap_vviz.png')
```

```{r}
plotEmbedding(adata$obsm['X_umap'],
              main = 'UMAP', xlab = "X", ylab = "Y")
```

```{r}
veloviz = buildVeloviz(
  curr = curr, proj = proj,
  normalize.depth = TRUE,
  use.ods.genes = TRUE,
  alpha = 0.05,
  pca = TRUE,
  nPCs = 20,
  center = TRUE,
  scale = TRUE,
  k = 20,
  similarity.threshold = 0.2,
  distance.weight = 1,
  distance.threshold = 1,
  weighted = TRUE,
  # seed = 0,
  verbose = FALSE
)
```

```{r}
emb.veloviz <- veloviz$fdg_coords
```

```{r}
plotEmbedding(emb.veloviz, colors = cell.cols[rownames(emb.veloviz)], main='VeloViz with dynamical velocity',
              xlab = "VeloViz X", ylab = "VeloViz Y",
              alpha = 0.8,
              cex.lab = 1.5)
```

```{r}
a.plot <- adata
```

```{r}
#remove unconnected cells from anndata object
connected.cells <-  rownames(emb.veloviz)
a.plot.vv <- a.plot[a.plot$obs_names$isin(connected.cells)]
a.plot.vv$obsm$update("X_veloviz1" = emb.veloviz)
```

```{r}
#get colors
plot.cols = unique(cell.cols)
c = unlist(sapply(c(1:length(col)), function(x) which(plot.cols == col[x])))
plot.cols = plot.cols[c]
```

```{r}
#plotting params
pt.size = 50
dnsty = 0.7
plt.size = c(3.5,3.5)


scv$pl$velocity_embedding_stream(a.plot.vv, basis='veloviz1',
                                 density = 0.8, cutoff_perc = 0, n_neighbors = 20L, title = "",legend_fontoutline = 2,
                                 size = 50, alpha = 0.6, legend_fontsize = 12, linewidth = 1.5,
                                 show = FALSE, figsize = c(5,5), palette = plot.cols) #, legend_loc = "lower left")
# plt$show()
plt$savefig("panc_dynamic_velocity.png")
```

```{r}
veloviz.nnGraph <- asNNGraph(veloviz) #converts veloviz igraph object to a format that UMAP understands
```

```{r}
set.seed(0)
emb.umapVelo <- uwot::umap(X = NULL, nn_method = veloviz.nnGraph, min_dist = 1)
rownames(emb.umapVelo) <- rownames(emb.veloviz)
plotEmbedding(emb.umapVelo, colors = cell.cols,
              main = 'VeloViz with UMAP', xlab = "X", ylab = "Y")
```

```{r}
par(mfrow = c(1,3))

# plotEmbedding(emb.umap, colors = cell.cols,
#               main = 'UMAP', xlab = "X", ylab = "Y")
plotEmbedding(emb.veloviz, colors = cell.cols,
              main = 'VeloViz with F-R', xlab = "X", ylab = "Y")
plotEmbedding(emb.umapVelo, colors = cell.cols,
              main = 'VeloViz with UMAP', xlab = "X", ylab = "Y")
```

